{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Processing process (python version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import graph_visualisation as gv\n",
    "\n",
    "from IPython.display import Image\n",
    "\n",
    "def save_png_and_svg_then_render_dot_graph(dot_source, filename_no_ext):\n",
    "    tmp_file = open(filename_no_ext, \"w\", encoding=\"utf-8\")\n",
    "    tmp_file.write(dot_source)\n",
    "    tmp_file.close()\n",
    "    \n",
    "    gv.render_graph(filename_no_ext, filetype='svg')\n",
    "    gv.render_graph(filename_no_ext, filetype='png')\n",
    "    \n",
    "    image = Image(filename=filename_no_ext + \".png\", embed=True, format='png')\n",
    "\n",
    "    return image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Graph "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Test data as pandas dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "data = pd.read_csv('./data/Test_RNAseq_Data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SRR7191189\n",
       "1    SRR7191190\n",
       "2    SRR7191191\n",
       "3    SRR7191192\n",
       "4    SRR7191193\n",
       "5    SRR7191194\n",
       "6    SRR7191195\n",
       "7    SRR7191196\n",
       "Name: Run, dtype: object"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Sample id\n",
    "data.Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!./fastqc "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download data by using fasterq-dump (sra)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tuturial (fasterq-dump): https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR7191189\n",
      "SRR7191190\n",
      "SRR7191191\n",
      "SRR7191192\n",
      "SRR7191193\n",
      "SRR7191194\n",
      "SRR7191195\n",
      "SRR7191196\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "for i in data.Run:\n",
    "    subprocess.run([\n",
    "        \"fasterq-dump\",\n",
    "    #     \"--split-3\",\n",
    "        i\n",
    "    ])\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TOBETESTED: TO parallelise this process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasterq-dump SRR6666666\n",
      "fasterq-dump SRR6666667\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "template = 'fasterq-dump {}'\n",
    "\n",
    "args = list(data.Run)\n",
    "# args = ['SRR6666666','SRR6666667']\n",
    "processes = []\n",
    "\n",
    "for arg in args:\n",
    "    command = template.format(*[arg])\n",
    "#     print(command)\n",
    "    process = subprocess.Popen(command, shell=True)\n",
    "    processes.append(process)\n",
    "\n",
    "# Collect statuses\n",
    "output = [p.wait() for p in processes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run FastQC to test the quality of those datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "            FastQC - A high throughput sequence QC analysis tool\r\n",
      "\r\n",
      "SYNOPSIS\r\n",
      "\r\n",
      "\tfastqc seqfile1 seqfile2 .. seqfileN\r\n",
      "\r\n",
      "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \r\n",
      "           [-c contaminant file] seqfile1 .. seqfileN\r\n",
      "\r\n",
      "DESCRIPTION\r\n",
      "\r\n",
      "    FastQC reads a set of sequence files and produces from each one a quality\r\n",
      "    control report consisting of a number of different modules, each one of \r\n",
      "    which will help to identify a different potential type of problem in your\r\n",
      "    data.\r\n",
      "    \r\n",
      "    If no files to process are specified on the command line then the program\r\n",
      "    will start as an interactive graphical application.  If files are provided\r\n",
      "    on the command line then the program will run with no user interaction\r\n",
      "    required.  In this mode it is suitable for inclusion into a standardised\r\n",
      "    analysis pipeline.\r\n",
      "    \r\n",
      "    The options for the program as as follows:\r\n",
      "    \r\n",
      "    -h --help       Print this help file and exit\r\n",
      "    \r\n",
      "    -v --version    Print the version of the program and exit\r\n",
      "    \r\n",
      "    -o --outdir     Create all output files in the specified output directory.\r\n",
      "                    Please note that this directory must exist as the program\r\n",
      "                    will not create it.  If this option is not set then the \r\n",
      "                    output file for each sequence file is created in the same\r\n",
      "                    directory as the sequence file which was processed.\r\n",
      "                    \r\n",
      "    --casava        Files come from raw casava output. Files in the same sample\r\n",
      "                    group (differing only by the group number) will be analysed\r\n",
      "                    as a set rather than individually. Sequences with the filter\r\n",
      "                    flag set in the header will be excluded from the analysis.\r\n",
      "                    Files must have the same names given to them by casava\r\n",
      "                    (including being gzipped and ending with .gz) otherwise they\r\n",
      "                    won't be grouped together correctly.\r\n",
      "                    \r\n",
      "    --nano          Files come from nanopore sequences and are in fast5 format. In\r\n",
      "                    this mode you can pass in directories to process and the program\r\n",
      "                    will take in all fast5 files within those directories and produce\r\n",
      "                    a single output file from the sequences found in all files.                    \r\n",
      "                    \r\n",
      "    --nofilter      If running with --casava then don't remove read flagged by\r\n",
      "                    casava as poor quality when performing the QC analysis.\r\n",
      "                   \r\n",
      "    --extract       If set then the zipped output file will be uncompressed in\r\n",
      "                    the same directory after it has been created.  By default\r\n",
      "                    this option will be set if fastqc is run in non-interactive\r\n",
      "                    mode.\r\n",
      "                    \r\n",
      "    -j --java       Provides the full path to the java binary you want to use to\r\n",
      "                    launch fastqc. If not supplied then java is assumed to be in\r\n",
      "                    your path.\r\n",
      "                   \r\n",
      "    --noextract     Do not uncompress the output file after creating it.  You\r\n",
      "                    should set this option if you do not wish to uncompress\r\n",
      "                    the output when running in non-interactive mode.\r\n",
      "                    \r\n",
      "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\r\n",
      "                    show data for every base in the read.  WARNING: Using this\r\n",
      "                    option will cause fastqc to crash and burn if you use it on\r\n",
      "                    really long reads, and your plots may end up a ridiculous size.\r\n",
      "                    You have been warned!\r\n",
      "                    \r\n",
      "    --min_length    Sets an artificial lower limit on the length of the sequence\r\n",
      "                    to be shown in the report.  As long as you set this to a value\r\n",
      "                    greater or equal to your longest read length then this will be\r\n",
      "                    the sequence length used to create your read groups.  This can\r\n",
      "                    be useful for making directly comaparable statistics from \r\n",
      "                    datasets with somewhat variable read lengths.\r\n",
      "                    \r\n",
      "    -f --format     Bypasses the normal sequence file format detection and\r\n",
      "                    forces the program to use the specified format.  Valid\r\n",
      "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\r\n",
      "                    \r\n",
      "    -t --threads    Specifies the number of files which can be processed\r\n",
      "                    simultaneously.  Each thread will be allocated 250MB of\r\n",
      "                    memory so you shouldn't run more threads than your\r\n",
      "                    available memory will cope with, and not more than\r\n",
      "                    6 threads on a 32 bit machine\r\n",
      "                  \r\n",
      "    -c              Specifies a non-default file which contains the list of\r\n",
      "    --contaminants  contaminants to screen overrepresented sequences against.\r\n",
      "                    The file must contain sets of named contaminants in the\r\n",
      "                    form name[tab]sequence.  Lines prefixed with a hash will\r\n",
      "                    be ignored.\r\n",
      "\r\n",
      "    -a              Specifies a non-default file which contains the list of\r\n",
      "    --adapters      adapter sequences which will be explicity searched against\r\n",
      "                    the library. The file must contain sets of named adapters\r\n",
      "                    in the form name[tab]sequence.  Lines prefixed with a hash\r\n",
      "                    will be ignored.\r\n",
      "                    \r\n",
      "    -l              Specifies a non-default file which contains a set of criteria\r\n",
      "    --limits        which will be used to determine the warn/error limits for the\r\n",
      "                    various modules.  This file can also be used to selectively \r\n",
      "                    remove some modules from the output all together.  The format\r\n",
      "                    needs to mirror the default limits.txt file found in the\r\n",
      "                    Configuration folder.\r\n",
      "                    \r\n",
      "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\r\n",
      "                    module. Specified Kmer length must be between 2 and 10. Default\r\n",
      "                    length is 7 if not specified.\r\n",
      "                    \r\n",
      "   -q --quiet       Supress all progress messages on stdout and only report errors.\r\n",
      "   \r\n",
      "   -d --dir         Selects a directory to be used for temporary files written when\r\n",
      "                    generating report images. Defaults to system temp directory if\r\n",
      "                    not specified.\r\n",
      "                    \r\n",
      "BUGS\r\n",
      "\r\n",
      "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\r\n",
      "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\r\n",
      "                   \r\n",
      "    \r\n"
     ]
    }
   ],
   "source": [
    "!fastqc --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So if your input has one file, the output should be ID.fastqc and ID.fastqc.zip. If you unzip the file, you will get a summary.txt file which indicates the results. Also, there is a results txt file for all the data that used for generating graphs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir testout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outputdir = 'testout/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastqfile_1 = \" \".join(list(data.Run + '_1' + '.fastq')) + \" -t \"+ str(len(data.Run))\n",
    "fastqfile_2 = \" \".join(list(data.Run + '_2' + '.fastq')) + \" -t \"+ str(len(data.Run))\n",
    "files = [fastqfile_1,fastqfile_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fastqc -o ./testout/ SRR7191189_2.fastq\n",
      "fastqc -o ./testout/ SRR7191190_2.fastq\n",
      "fastqc -o ./testout/ SRR7191191_2.fastq\n",
      "fastqc -o ./testout/ SRR7191192_2.fastq\n",
      "fastqc -o ./testout/ SRR7191193_2.fastq\n",
      "fastqc -o ./testout/ SRR7191194_2.fastq\n",
      "fastqc -o ./testout/ SRR7191195_2.fastq\n",
      "fastqc -o ./testout/ SRR7191196_2.fastq\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "template = 'fastqc -o ./testout/ {}'\n",
    "\n",
    "args = list(data.Run + '_2' + '.fastq') # run for _1 as well\n",
    "\n",
    "processes = []\n",
    "\n",
    "for arg in args:\n",
    "    command = template.format(*[arg])\n",
    "#     print(command)\n",
    "    process = subprocess.Popen(command, shell=True)\n",
    "    processes.append(process)\n",
    "\n",
    "# Collect statuses\n",
    "output = [p.wait() for p in processes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It takes forever, personally think it might not be a good choice to do this in python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"700\"\n",
       "            height=\"600\"\n",
       "            src=\"testout/SRR7191189_2_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7feac2c1db10>"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "\n",
    "IFrame(src='testout/SRR7191189_2_fastqc.html', width=700, height=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get all the fastQC output filenames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "onlyfiles = [f for f in listdir('testout') if isfile(join('testout', f))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "identify = []\n",
    "\n",
    "for i in onlyfiles:\n",
    "    if i[-1] == 'p':\n",
    "        subprocess.run([\n",
    "            \"unzip\",'./testout/'+ i, '-d', './summary'\n",
    "        ])\n",
    "        identify.append(i[0:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 609,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##FastQC\t0.11.9\r\n",
      ">>Basic Statistics\tpass\r\n",
      "#Measure\tValue\r\n",
      "Filename\tSRR7191189_1.fastq\r\n",
      "File type\tConventional base calls\r\n",
      "Encoding\tSanger / Illumina 1.9\r\n",
      "Total Sequences\t68452760\r\n",
      "Sequences flagged as poor quality\t0\r\n",
      "Sequence length\t25-100\r\n",
      "%GC\t50\r\n"
     ]
    }
   ],
   "source": [
    "!head SRR7191189_1_fastqc/fastqc_data.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_overseq(filename, withoutAT = True):\n",
    "    \n",
    "    f = open(filename,'r')\n",
    "    txtread = f.read()\n",
    "    sequence = []\n",
    "    in_code_section = False\n",
    "    # id filename.split('/')[2]\n",
    "    \n",
    "    for line in txtread.split('\\n'):\n",
    "        line = line.strip()\n",
    "        \n",
    "        if line == '#Sequence\tCount\tPercentage\tPossible Source':\n",
    "            in_code_section = True\n",
    "            continue\n",
    "        \n",
    "        if line.startswith(\">>END_MODULE\"):\n",
    "            in_code_section = in_code_section*-1\n",
    "            \n",
    "        if in_code_section and not line.startswith(\">>END_MODULE\"):\n",
    "            line.split('\\t')\n",
    "        \n",
    "            tmp_seq = line.split('\\t')[0]\n",
    "            \n",
    "            if (tmp_seq == 'T'*50 or tmp_seq == 'A'*50) and withoutAT:\n",
    "                continue\n",
    "            else:\n",
    "                sequence.append(tmp_seq)\n",
    "        \n",
    "        if in_code_section == -1:\n",
    "            return sequence\n",
    "\n",
    "    return sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "Basic_stat = None\n",
    "df = []\n",
    "\n",
    "\n",
    "for i in identify:\n",
    "\n",
    "    filename = './summary/' + i + '/summary.txt'\n",
    "    tmp = pd.read_csv(filename, delimiter='\\t')\n",
    "    new_row = [j for j in tmp.PASS]\n",
    "    new_row.insert(0,i[0:-7])\n",
    "    df.append(new_row)\n",
    "    \n",
    "    if Basic_stat is None:\n",
    "\n",
    "        Basic_stat = list(tmp.iloc[:,1])\n",
    "        Basic_stat.insert(0,'id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./summary/SRR7191194_2_fastqc/fastqc_data.txt'"
      ]
     },
     "execution_count": 502,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sequences = []\n",
    "\n",
    "for i in identify:\n",
    "\n",
    "    filename = './summary/' + i + '/fastqc_data.txt'\n",
    "    sequences.append(find_overseq(filename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 541,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[],\n",
       " [],\n",
       " [],\n",
       " ['GCACACGTCTGAACTCCAGTCACCCTATCCTATATCTCGTATGCCGTCTT'],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " ['GCACACGTCTGAACTCCAGTCACTATAGCCTATATCTCGTATGCCGTCTT'],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " []]"
      ]
     },
     "execution_count": 541,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(df,columns=Basic_stat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 542,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['sequences_withoutAT'] = sequences "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "metadata": {},
   "outputs": [],
   "source": [
    "# onlyfiles = [f for f in listdir('testout') if isfile(join('testout', f))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_adapter(filename):\n",
    "    \n",
    "    with open(filename,'r') as f:\n",
    "        txtread = f.read()\n",
    "    sequence = []\n",
    "    in_code_section = False\n",
    "    # id filename.split('/')[2]\n",
    "    \n",
    "    for line in txtread.split('\\n'):\n",
    "        line = line.strip()\n",
    "        \n",
    "        if line == \"#Position\tIllumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tSOLID Small RNA Adapter\":\n",
    "            in_code_section = True\n",
    "            continue\n",
    "        \n",
    "        if line.startswith(\">>END_MODULE\"):\n",
    "            in_code_section = in_code_section*-1\n",
    "            \n",
    "        if in_code_section and not line.startswith(\">>END_MODULE\"):\n",
    "            line.split('\\t')\n",
    "            sequence.append(line.split('\\t'))\n",
    "        \n",
    "        if in_code_section == -1:\n",
    "            return sequence[-1][1:6]\n",
    "\n",
    "    return 'Cannot find anything'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "metadata": {},
   "outputs": [],
   "source": [
    "adapters = []\n",
    "\n",
    "for i in identify:\n",
    "\n",
    "    filename = './summary/' + i + '/fastqc_data.txt'\n",
    "    adapter = find_adapter(filename)\n",
    "    adapters.append([float(i)>1 for i in adapter])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 569,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['adapters'] = adapters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 570,
   "metadata": {},
   "outputs": [],
   "source": [
    "adapter_names = \"Illumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tSOLID Small RNA Adapter\".split('\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 571,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Illumina Universal Adapter',\n",
       " \"Illumina Small RNA 3' Adapter\",\n",
       " \"Illumina Small RNA 5' Adapter\",\n",
       " 'Nextera Transposase Sequence',\n",
       " 'SOLID Small RNA Adapter']"
      ]
     },
     "execution_count": 571,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adapter_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set adapter dictionary (can be found in fastqc adapter_list.txt) could be generated automatically\n",
    "dictionary = {'Illumina Universal Adapter':'AGATCGGAAGAG',\n",
    "              \"Illumina Small RNA 3' Adapter\":'TGGAATTCTCGG',\n",
    "              \"Illumina Small RNA 5' Adapter\":'GATCGTCGGACT',\n",
    "              'Nextera Transposase Sequence':'CTGTCTCTTATA',\n",
    "              'SOLID Small RNA Adapter':'CGCCTTGGCCGT',\n",
    "             }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 600,
   "metadata": {},
   "outputs": [],
   "source": [
    "adapter_seqs = []\n",
    "\n",
    "for adapter_list in adapters:\n",
    "   \n",
    "    adapter_seq = []\n",
    "    \n",
    "    for i,adapter_name in zip(adapter_list,adapter_names):\n",
    "        \n",
    "        if i == True:\n",
    "            adapter_seq.append(dictionary[adapter_name])\n",
    "            \n",
    "    adapter_seqs.append(adapter_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['adapter_seqs'] = adapter_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run cutadapter to cut off the low quality data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir cutpair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Per base sequence quality</th>\n",
       "      <th>Per sequence quality scores</th>\n",
       "      <th>Per base sequence content</th>\n",
       "      <th>Per sequence GC content</th>\n",
       "      <th>Per base N content</th>\n",
       "      <th>Sequence Length Distribution</th>\n",
       "      <th>Sequence Duplication Levels</th>\n",
       "      <th>Overrepresented sequences</th>\n",
       "      <th>Adapter Content</th>\n",
       "      <th>sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR7191194_2</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>WARN</td>\n",
       "      <td>FAIL</td>\n",
       "      <td>WARN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>[TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id Per base sequence quality Per sequence quality scores  \\\n",
       "15  SRR7191194_2                      PASS                        PASS   \n",
       "\n",
       "   Per base sequence content Per sequence GC content Per base N content  \\\n",
       "15                      PASS                    PASS               PASS   \n",
       "\n",
       "   Sequence Length Distribution Sequence Duplication Levels  \\\n",
       "15                         WARN                        FAIL   \n",
       "\n",
       "   Overrepresented sequences Adapter Content  \\\n",
       "15                      WARN            PASS   \n",
       "\n",
       "                                            sequences  \n",
       "15  [TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...  "
      ]
     },
     "execution_count": 384,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input1 = i + '_1.fastq'\n",
    "input2 = i + '_2.fastq'\n",
    "\n",
    "df.loc[df.id == i[0:12]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parameter_generation(sequence1, parameter = '-a'):\n",
    "    adapters1 = []\n",
    "    if len(sequence1.iloc[0]) is not 0:\n",
    "        for j in sequence1.iloc[0]:\n",
    "            adapters1.append(parameter)\n",
    "            adapters1.append(j)\n",
    "    return adapters1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt -q 20 -a GCACACGTCTGAACTCCAGTCACTATAGCCTATATCTCGTATGCCGTCTT -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191189_1.fastq -p cutpair/SRR7191189_2.fastq -j 80 SRR7191189_1.fastq SRR7191189_2.fastq\n",
      "cutadapt -q 20 -a GCACACGTCTGAACTCCAGTCACCCTATCCTATATCTCGTATGCCGTCTT -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191190_1.fastq -p cutpair/SRR7191190_2.fastq -j 80 SRR7191190_1.fastq SRR7191190_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191191_1.fastq -p cutpair/SRR7191191_2.fastq -j 80 SRR7191191_1.fastq SRR7191191_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191192_1.fastq -p cutpair/SRR7191192_2.fastq -j 80 SRR7191192_1.fastq SRR7191192_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191193_1.fastq -p cutpair/SRR7191193_2.fastq -j 80 SRR7191193_1.fastq SRR7191193_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191194_1.fastq -p cutpair/SRR7191194_2.fastq -j 80 SRR7191194_1.fastq SRR7191194_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191195_1.fastq -p cutpair/SRR7191195_2.fastq -j 80 SRR7191195_1.fastq SRR7191195_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191196_1.fastq -p cutpair/SRR7191196_2.fastq -j 80 SRR7191196_1.fastq SRR7191196_2.fastq\n"
     ]
    }
   ],
   "source": [
    "tmpdir = 'cutpair/'\n",
    "\n",
    "for i in data.Run:\n",
    "    \n",
    "    input1 = i + '_1'\n",
    "    input2 = i + '_2'   \n",
    "    input_dir1 = i + '_1.fastq'\n",
    "    input_dir2 = i + '_2.fastq'\n",
    "    output_dir1 = tmpdir + i + '_1.fastq'\n",
    "    output_dir2 = tmpdir + i + '_2.fastq'\n",
    "    \n",
    "    sequence1 = df.loc[df.id == input1].sequences_withoutAT # change here for AAA\n",
    "    sequence2 = df.loc[df.id == input2].sequences_withoutAT\n",
    "    \n",
    "    adapter_seq1 = df.loc[df.id == input1].adapter_seqs\n",
    "    adapter_seq2 = df.loc[df.id == input1].adapter_seqs\n",
    "    \n",
    "    # -a and -A for input1 and input2\n",
    "    over_seq1 = parameter_generation(sequence1)\n",
    "    over_seq2 = parameter_generation(sequence2,'-A')\n",
    "\n",
    "    # -a and -A for adapters\n",
    "    adapter1 = parameter_generation(adapter_seq1)\n",
    "    adapter2 = parameter_generation(adapter_seq2,'-A')\n",
    "    \n",
    "    # command generation\n",
    "    command = ['cutadapt', '-q', '20'] # cut down the low quality reads\n",
    "    command.extend(over_seq1)\n",
    "    command.extend(over_seq2)\n",
    "    command.extend(adapter1)\n",
    "    command.extend(adapter2)\n",
    "\n",
    "    command.extend(['-o', output_dir1, # output_dir for input1\n",
    "                    '-p', output_dir2, # output_dir for input2\n",
    "                    '-j', '80',\n",
    "                    input_dir1,\n",
    "                    input_dir2])\n",
    "    print(\" \".join(command))\n",
    "\n",
    "    subprocess.run(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 2.8 with Python 3.7.6\n",
      "Command line parameters: -q 20 -o cutpair/out.1.fastq -p cutpair/out.2.fastq SRR7191192_1.fastq SRR7191192_2.fastq\n",
      "Processing reads on 1 core in paired-end mode ...\n",
      "[------->8   ] 00:12:10    77,263,058 reads  @      9.5 µs/read;   6.34 M reads/minute\n",
      "Finished in 731.00 s (9 us/read; 6.34 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total read pairs processed:         77,263,058\n",
      "  Read 1 with adapter:                       0 (0.0%)\n",
      "  Read 2 with adapter:                       0 (0.0%)\n",
      "Pairs written (passing filters):    77,263,058 (100.0%)\n",
      "\n",
      "Total basepairs processed: 15,452,610,605 bp\n",
      "  Read 1: 7,726,305,531 bp\n",
      "  Read 2: 7,726,305,074 bp\n",
      "Quality-trimmed:             201,314,526 bp (1.3%)\n",
      "  Read 1:    41,944,144 bp\n",
      "  Read 2:   159,370,382 bp\n",
      "Total written (filtered):  15,251,296,079 bp (98.7%)\n",
      "  Read 1: 7,684,361,387 bp\n",
      "  Read 2: 7,566,934,692 bp\n"
     ]
    }
   ],
   "source": [
    "!cutadapt -q 20 -o cutpair/out.1.fastq -p cutpair/out.2.fastq SRR7191192_1.fastq SRR7191192_2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 2.8 with Python 3.7.6\n",
      "Command line parameters: -q 20 -o cut/SRR7191192_2.fastq SRR7191192_2.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[---->8      ] 00:05:06    77,263,058 reads  @      4.0 µs/read;  15.12 M reads/minute\n",
      "Finished in 306.68 s (4 us/read; 15.12 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              77,263,058\n",
      "Reads with adapters:                         0 (0.0%)\n",
      "Reads written (passing filters):    77,263,058 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,726,305,074 bp\n",
      "Quality-trimmed:             159,370,382 bp (2.1%)\n",
      "Total written (filtered):  7,566,934,692 bp (97.9%)\n"
     ]
    }
   ],
   "source": [
    "!cutadapt -q 20 -o cut/SRR7191192_2.fastq SRR7191192_2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "args_1 = list(data.Run + '_1' + '.fastq')\n",
    "args_2 = list(data.Run + '_2' + '.fastq')\n",
    "args_1 = args_1[0:-1]\n",
    "args_2 = args_2[0:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR --runThreadN 10 --genomeDir ./genome --readFilesIn ./data/SRR7191196_1.fastq ./data/SRR7191196_2.fastq --outFileNamePrefix ./test_sample/ --outSAMtype BAM SortedByCoordinate --sjdbGTFfile gencode.v33.primary_assembly.annotation.gtf --sjdbOverhang 100 --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomedir = '/mnt/Tank/junfan/project/test1/genome'\n",
    "GTFdir = '/mnt/Tank/junfan/project/test1/gencode.v33.primary_assembly.annotation.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir STAROUTPUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SRR7191189_1.fastq',\n",
       " 'SRR7191190_1.fastq',\n",
       " 'SRR7191191_1.fastq',\n",
       " 'SRR7191192_1.fastq',\n",
       " 'SRR7191193_1.fastq',\n",
       " 'SRR7191194_1.fastq',\n",
       " 'SRR7191195_1.fastq']"
      ]
     },
     "execution_count": 498,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "template = 'STAR --runThreadN 80 --genomeDir ' + genomedir + ' --readFilesIn {} {} --outFileNamePrefix ./{} --outSAMtype BAM SortedByCoordinate --sjdbGTFfile '+ GTFdir + ' --sjdbOverhang 100 --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded'\n",
    "\n",
    "parent_dir = \"/mnt/Tank/junfan/project/test\"\n",
    "\n",
    "commands = []\n",
    "\n",
    "readfiledir = './'\n",
    "\n",
    "# subprocess.run(['ulimit', '-n', '4096'])\n",
    "\n",
    "with open('star-command.sh','w+') as f:\n",
    "    \n",
    "    f.write('# set limit\\n')\n",
    "    f.write('ulimit -n 4096\\n')\n",
    "    \n",
    "    for input1, input2 in zip(args_1,args_2):\n",
    "\n",
    "        # create folders\n",
    "\n",
    "    #     output_folder = 'STAROUTPUT/' + input1[0:10]\n",
    "    #     path = os.path.join(parent_dir, output_folder)\n",
    "    #     try: \n",
    "    #         os.mkdir(path) \n",
    "    #     except OSError as error: \n",
    "    #         print(error)     \n",
    "\n",
    "        command = template.format(*[readfiledir+input1, readfiledir+input2, output_folder])\n",
    "        f.write('\\n\\n# id:'+ input1[0:10] + '\\n\\n')\n",
    "        f.write(command)\n",
    "#         subprocess.run(command.split(' '))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'STAR --runThreadN 80 --genomeDir /mnt/Tank/junfan/project/test1/genome --readFilesIn ./SRR7191195_1.fastq ./SRR7191195_2.fastq --outFileNamePrefix ./STAROUTPUT/SRR7191195 --outSAMtype BAM SortedByCoordinate --sjdbGTFfile /mnt/Tank/junfan/project/test1/gencode.v33.primary_assembly.annotation.gtf --sjdbOverhang 100 --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded'"
      ]
     },
     "execution_count": 504,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Jan 24 10:41:36 ..... started STAR run\n",
      "Jan 24 10:41:36 ..... loading genome\n",
      "Jan 24 10:41:54 ..... processing annotations GTF\n",
      "Jan 24 10:42:06 ..... inserting junctions into the genome indices\n",
      "^C\n"
     ]
    }
   ],
   "source": [
    "!bash star-command.sh "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "template = 'STAR --runThreadN 10 --genomeDir ' + genomedir + ' --readFilesIn {} {} --outFileNamePrefix ./{} --outSAMtype BAM SortedByCoordinate --sjdbGTFfile '+ GTFdir + ' --sjdbOverhang 100 --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded'\n",
    "\n",
    "parent_dir = \"/mnt/Tank/junfan/project/test\"\n",
    "\n",
    "processes = []\n",
    "\n",
    "for input1, input2 in zip(args_1,args_2):\n",
    "    \n",
    "    # create folders\n",
    "    \n",
    "#     output_folder = 'STAROUTPUT/' + input1[0:10]\n",
    "#     path = os.path.join(parent_dir, output_folder)\n",
    "#     try: \n",
    "#         os.mkdir(path) \n",
    "#     except OSError as error: \n",
    "#         print(error)     \n",
    "    \n",
    "    command = template.format(*[input1, input2, output_folder])\n",
    "    process = subprocess.Popen(command, shell=True)\n",
    "    processes.append(process)\n",
    "\n",
    "# Collect statuses\n",
    "output = [p.wait() for p in processes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counts to Different Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are two tools to use: **edgeR** and **DESeq2**\n",
    "```{R}\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"edgeR\")\n",
    "```\n",
    "\n",
    "In order to use edgeR, you will also need to install limma.\n",
    "```\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"limma\")\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(limma)\n",
    "library(edgeR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR --runThreadN 79 --runMode genomeGenerate --genomeDir ./ \\--genomeFastaFiles ./Homo_sapiens.GRCh38.dna.primary_assembly.fa \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR \\--runThreadN 12 --genomeDir ~/star/genome/ \\--sjdbGTFfile ~/star/Homo_sapiens.GRCh38.79.gtf --sjdbOverhang 100 \\--readFilesIn SRR7191196_1.fastq SRR7191196_2.fastq \\--readFilesCommand zcat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR --runThreadN 20 --readFilesIn tq --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /home/fanyc/RNA-seq/STAR/23 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR --runThreadN 20 --readFilesIn SRR7191196_1.fastq SRR7191196_2.fastq --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix ./output --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasterq-dump --split-3 SRR7191190"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "process = subprocess.Popen(['echo', 'More output'],\n",
    "                     stdout=subprocess.PIPE, \n",
    "                     stderr=subprocess.PIPE)\n",
    "stdout, stderr = process.communicate()\n",
    "stdout, stderr"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
