{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Processing process (python version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import graph_visualisation as gv\n",
    "\n",
    "from IPython.display import Image\n",
    "\n",
    "def save_png_and_svg_then_render_dot_graph(dot_source, filename_no_ext):\n",
    "    tmp_file = open(filename_no_ext, \"w\", encoding=\"utf-8\")\n",
    "    tmp_file.write(dot_source)\n",
    "    tmp_file.close()\n",
    "    \n",
    "    gv.render_graph(filename_no_ext, filetype='svg')\n",
    "    gv.render_graph(filename_no_ext, filetype='png')\n",
    "    \n",
    "    image = Image(filename=filename_no_ext + \".png\", embed=True, format='png')\n",
    "\n",
    "    return image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Graph process"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Test data as pandas dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "data = pd.read_csv('./data/Test_RNAseq_Data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SRR7191189\n",
       "1    SRR7191190\n",
       "2    SRR7191191\n",
       "3    SRR7191192\n",
       "4    SRR7191193\n",
       "5    SRR7191194\n",
       "6    SRR7191195\n",
       "7    SRR7191196\n",
       "Name: Run, dtype: object"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Sample id\n",
    "data.Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!./fastqc "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download data by using fasterq-dump (sra)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tuturial (fasterq-dump): https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR7191189\n",
      "SRR7191190\n",
      "SRR7191191\n",
      "SRR7191192\n",
      "SRR7191193\n",
      "SRR7191194\n",
      "SRR7191195\n",
      "SRR7191196\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "for i in data.Run:\n",
    "    subprocess.run([\n",
    "        \"fasterq-dump\",\n",
    "    #     \"--split-3\",\n",
    "        i\n",
    "    ])\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TOBETESTED: TO parallelise this process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasterq-dump SRR6666666\n",
      "fasterq-dump SRR6666667\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "template = 'fasterq-dump {}'\n",
    "\n",
    "args = list(data.Run)\n",
    "# args = ['SRR6666666','SRR6666667']\n",
    "processes = []\n",
    "\n",
    "for arg in args:\n",
    "    command = template.format(*[arg])\n",
    "#     print(command)\n",
    "    process = subprocess.Popen(command, shell=True)\n",
    "    processes.append(process)\n",
    "\n",
    "# Collect statuses\n",
    "output = [p.wait() for p in processes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run FastQC to test the quality of those datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "            FastQC - A high throughput sequence QC analysis tool\r\n",
      "\r\n",
      "SYNOPSIS\r\n",
      "\r\n",
      "\tfastqc seqfile1 seqfile2 .. seqfileN\r\n",
      "\r\n",
      "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \r\n",
      "           [-c contaminant file] seqfile1 .. seqfileN\r\n",
      "\r\n",
      "DESCRIPTION\r\n",
      "\r\n",
      "    FastQC reads a set of sequence files and produces from each one a quality\r\n",
      "    control report consisting of a number of different modules, each one of \r\n",
      "    which will help to identify a different potential type of problem in your\r\n",
      "    data.\r\n",
      "    \r\n",
      "    If no files to process are specified on the command line then the program\r\n",
      "    will start as an interactive graphical application.  If files are provided\r\n",
      "    on the command line then the program will run with no user interaction\r\n",
      "    required.  In this mode it is suitable for inclusion into a standardised\r\n",
      "    analysis pipeline.\r\n",
      "    \r\n",
      "    The options for the program as as follows:\r\n",
      "    \r\n",
      "    -h --help       Print this help file and exit\r\n",
      "    \r\n",
      "    -v --version    Print the version of the program and exit\r\n",
      "    \r\n",
      "    -o --outdir     Create all output files in the specified output directory.\r\n",
      "                    Please note that this directory must exist as the program\r\n",
      "                    will not create it.  If this option is not set then the \r\n",
      "                    output file for each sequence file is created in the same\r\n",
      "                    directory as the sequence file which was processed.\r\n",
      "                    \r\n",
      "    --casava        Files come from raw casava output. Files in the same sample\r\n",
      "                    group (differing only by the group number) will be analysed\r\n",
      "                    as a set rather than individually. Sequences with the filter\r\n",
      "                    flag set in the header will be excluded from the analysis.\r\n",
      "                    Files must have the same names given to them by casava\r\n",
      "                    (including being gzipped and ending with .gz) otherwise they\r\n",
      "                    won't be grouped together correctly.\r\n",
      "                    \r\n",
      "    --nano          Files come from nanopore sequences and are in fast5 format. In\r\n",
      "                    this mode you can pass in directories to process and the program\r\n",
      "                    will take in all fast5 files within those directories and produce\r\n",
      "                    a single output file from the sequences found in all files.                    \r\n",
      "                    \r\n",
      "    --nofilter      If running with --casava then don't remove read flagged by\r\n",
      "                    casava as poor quality when performing the QC analysis.\r\n",
      "                   \r\n",
      "    --extract       If set then the zipped output file will be uncompressed in\r\n",
      "                    the same directory after it has been created.  By default\r\n",
      "                    this option will be set if fastqc is run in non-interactive\r\n",
      "                    mode.\r\n",
      "                    \r\n",
      "    -j --java       Provides the full path to the java binary you want to use to\r\n",
      "                    launch fastqc. If not supplied then java is assumed to be in\r\n",
      "                    your path.\r\n",
      "                   \r\n",
      "    --noextract     Do not uncompress the output file after creating it.  You\r\n",
      "                    should set this option if you do not wish to uncompress\r\n",
      "                    the output when running in non-interactive mode.\r\n",
      "                    \r\n",
      "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\r\n",
      "                    show data for every base in the read.  WARNING: Using this\r\n",
      "                    option will cause fastqc to crash and burn if you use it on\r\n",
      "                    really long reads, and your plots may end up a ridiculous size.\r\n",
      "                    You have been warned!\r\n",
      "                    \r\n",
      "    --min_length    Sets an artificial lower limit on the length of the sequence\r\n",
      "                    to be shown in the report.  As long as you set this to a value\r\n",
      "                    greater or equal to your longest read length then this will be\r\n",
      "                    the sequence length used to create your read groups.  This can\r\n",
      "                    be useful for making directly comaparable statistics from \r\n",
      "                    datasets with somewhat variable read lengths.\r\n",
      "                    \r\n",
      "    -f --format     Bypasses the normal sequence file format detection and\r\n",
      "                    forces the program to use the specified format.  Valid\r\n",
      "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\r\n",
      "                    \r\n",
      "    -t --threads    Specifies the number of files which can be processed\r\n",
      "                    simultaneously.  Each thread will be allocated 250MB of\r\n",
      "                    memory so you shouldn't run more threads than your\r\n",
      "                    available memory will cope with, and not more than\r\n",
      "                    6 threads on a 32 bit machine\r\n",
      "                  \r\n",
      "    -c              Specifies a non-default file which contains the list of\r\n",
      "    --contaminants  contaminants to screen overrepresented sequences against.\r\n",
      "                    The file must contain sets of named contaminants in the\r\n",
      "                    form name[tab]sequence.  Lines prefixed with a hash will\r\n",
      "                    be ignored.\r\n",
      "\r\n",
      "    -a              Specifies a non-default file which contains the list of\r\n",
      "    --adapters      adapter sequences which will be explicity searched against\r\n",
      "                    the library. The file must contain sets of named adapters\r\n",
      "                    in the form name[tab]sequence.  Lines prefixed with a hash\r\n",
      "                    will be ignored.\r\n",
      "                    \r\n",
      "    -l              Specifies a non-default file which contains a set of criteria\r\n",
      "    --limits        which will be used to determine the warn/error limits for the\r\n",
      "                    various modules.  This file can also be used to selectively \r\n",
      "                    remove some modules from the output all together.  The format\r\n",
      "                    needs to mirror the default limits.txt file found in the\r\n",
      "                    Configuration folder.\r\n",
      "                    \r\n",
      "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\r\n",
      "                    module. Specified Kmer length must be between 2 and 10. Default\r\n",
      "                    length is 7 if not specified.\r\n",
      "                    \r\n",
      "   -q --quiet       Supress all progress messages on stdout and only report errors.\r\n",
      "   \r\n",
      "   -d --dir         Selects a directory to be used for temporary files written when\r\n",
      "                    generating report images. Defaults to system temp directory if\r\n",
      "                    not specified.\r\n",
      "                    \r\n",
      "BUGS\r\n",
      "\r\n",
      "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\r\n",
      "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\r\n",
      "                   \r\n",
      "    \r\n"
     ]
    }
   ],
   "source": [
    "!fastqc --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So if your input has one file, the output should be ID.fastqc and ID.fastqc.zip. If you unzip the file, you will get a summary.txt file which indicates the results. Also, there is a results txt file for all the data that used for generating graphs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir testout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outputdir = 'testout/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastqfile_1 = \" \".join(list(data.Run + '_1' + '.fastq')) + \" -t \"+ str(len(data.Run))\n",
    "fastqfile_2 = \" \".join(list(data.Run + '_2' + '.fastq')) + \" -t \"+ str(len(data.Run))\n",
    "files = [fastqfile_1,fastqfile_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fastqc -o ./testout/ SRR7191189_2.fastq\n",
      "fastqc -o ./testout/ SRR7191190_2.fastq\n",
      "fastqc -o ./testout/ SRR7191191_2.fastq\n",
      "fastqc -o ./testout/ SRR7191192_2.fastq\n",
      "fastqc -o ./testout/ SRR7191193_2.fastq\n",
      "fastqc -o ./testout/ SRR7191194_2.fastq\n",
      "fastqc -o ./testout/ SRR7191195_2.fastq\n",
      "fastqc -o ./testout/ SRR7191196_2.fastq\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "template = 'fastqc -o ./testout/ {}'\n",
    "\n",
    "args = list(data.Run + '_2' + '.fastq') # run for _1 as well\n",
    "\n",
    "processes = []\n",
    "\n",
    "for arg in args:\n",
    "    command = template.format(*[arg])\n",
    "#     print(command)\n",
    "    process = subprocess.Popen(command, shell=True)\n",
    "    processes.append(process)\n",
    "\n",
    "# Collect statuses\n",
    "output = [p.wait() for p in processes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It takes forever, personally think it might not be a good choice to do this in python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"700\"\n",
       "            height=\"600\"\n",
       "            src=\"testout/SRR7191189_2_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7feac2c1db10>"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "\n",
    "IFrame(src='testout/SRR7191189_2_fastqc.html', width=700, height=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get all the fastQC output filenames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "onlyfiles = [f for f in listdir('testout') if isfile(join('testout', f))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "identify = []\n",
    "\n",
    "for i in onlyfiles:\n",
    "    if i[-1] == 'p':\n",
    "        subprocess.run([\n",
    "            \"unzip\",'./testout/'+ i, '-d', './summary'\n",
    "        ])\n",
    "        identify.append(i[0:-4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##FastQC\t0.11.9\r\n",
      ">>Basic Statistics\tpass\r\n",
      "#Measure\tValue\r\n",
      "Filename\tSRR7191189_1.fastq\r\n",
      "File type\tConventional base calls\r\n",
      "Encoding\tSanger / Illumina 1.9\r\n",
      "Total Sequences\t68452760\r\n",
      "Sequences flagged as poor quality\t0\r\n",
      "Sequence length\t25-100\r\n",
      "%GC\t50\r\n",
      ">>END_MODULE\r\n",
      ">>Per base sequence quality\tpass\r\n",
      "#Base\tMean\tMedian\tLower Quartile\tUpper Quartile\t10th Percentile\t90th Percentile\r\n",
      "1\t40.059466440798005\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "2\t40.03664943531861\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "3\t40.06407043339085\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "4\t40.04433609397196\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "5\t40.034695708398026\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "6\t40.04169625885063\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "7\t40.02143649138471\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "8\t40.01407240555385\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "9\t40.034389409572384\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "10-11\t40.012003343619746\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "12-13\t40.001757504007145\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "14-15\t39.97554508393818\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "16-17\t39.903069795870906\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "18-19\t39.86958782231717\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "20-21\t39.77426764822923\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "22-23\t39.795784340908966\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "24-25\t39.7562212538983\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "26-27\t39.71452414984542\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "28-29\t39.6572751189554\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "30-31\t39.61790045943657\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "32-33\t39.57133397855692\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "34-35\t39.536157075455705\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "36-37\t39.4963000321521\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "38-39\t39.479655494372565\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "40-41\t39.44409864795213\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "42-43\t39.41974693231481\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "44-45\t39.388958791928225\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "46-47\t39.35126229545125\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "48-49\t39.35584692472709\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "50-51\t39.29899084849092\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "52-53\t39.2724966332253\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "54-55\t39.23284215854923\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "56-57\t39.20062017519412\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "58-59\t39.159041256397074\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "60-61\t39.157631985120716\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "62-63\t39.091872600187855\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "64-65\t39.01048731362948\t41.0\t41.0\t41.0\t37.0\t41.0\r\n",
      "66-67\t38.608727210595504\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "68-69\t38.862148453515836\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "70-71\t38.90985361962208\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "72-73\t38.8983066546614\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "74-75\t38.88912695008775\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "76-77\t38.84179881706146\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "78-79\t38.809000689926286\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "80-81\t38.78377040999024\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "82-83\t38.735023049182885\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "84-85\t38.6895876005102\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "86-87\t38.6364256253715\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "88-89\t38.60050895397164\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "90-91\t38.556312951450074\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "92-93\t38.505698222330054\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "94-95\t38.44113082553642\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "96-97\t38.3890069605967\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "98-99\t38.33486899840054\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      "100\t38.280849351389115\t41.0\t41.0\t41.0\t32.0\t41.0\r\n",
      ">>END_MODULE\r\n",
      ">>Per sequence quality scores\tpass\r\n",
      "#Quality\tCount\r\n",
      "2\t3.0\r\n",
      "3\t0.0\r\n",
      "4\t0.0\r\n",
      "5\t0.0\r\n",
      "6\t0.0\r\n",
      "7\t0.0\r\n",
      "8\t0.0\r\n",
      "9\t0.0\r\n",
      "10\t0.0\r\n",
      "11\t0.0\r\n",
      "12\t30.0\r\n",
      "13\t54.0\r\n",
      "14\t105.0\r\n",
      "15\t352.0\r\n",
      "16\t1096.0\r\n",
      "17\t2178.0\r\n",
      "18\t4812.0\r\n",
      "19\t12346.0\r\n",
      "20\t32703.0\r\n",
      "21\t74463.0\r\n",
      "22\t133170.0\r\n",
      "23\t181389.0\r\n",
      "24\t211889.0\r\n",
      "25\t236333.0\r\n",
      "26\t267669.0\r\n",
      "27\t315704.0\r\n",
      "28\t386394.0\r\n",
      "29\t480375.0\r\n",
      "30\t636750.0\r\n",
      "31\t767751.0\r\n",
      "32\t850009.0\r\n",
      "33\t1009467.0\r\n",
      "34\t1203224.0\r\n",
      "35\t1449386.0\r\n",
      "36\t1782178.0\r\n",
      "37\t2283277.0\r\n",
      "38\t3188459.0\r\n",
      "39\t5444341.0\r\n",
      "40\t3.2946464E7\r\n",
      "41\t1.4550389E7\r\n",
      ">>END_MODULE\r\n",
      ">>Per base sequence content\tpass\r\n",
      "#Base\tG\tA\tT\tC\r\n",
      "1\t27.529241809792993\t24.399676115309717\t23.263283610329967\t24.80779846456732\r\n",
      "2\t26.11374804629681\t21.346846020392174\t27.582290229043544\t24.957115704267473\r\n",
      "3\t25.34937053590818\t23.013448414558287\t27.500059785762016\t24.137121263771526\r\n",
      "4\t24.532021405653538\t23.797535284079206\t27.96144646153551\t23.708996848731747\r\n",
      "5\t24.971055585094987\t23.770100304360277\t27.419259387901647\t23.839584722643092\r\n",
      "6\t25.452935770938762\t23.39245615720461\t26.978794073903707\t24.175813997952925\r\n",
      "7\t25.293714612546577\t23.852358495947794\t27.128302516726976\t23.725624374778654\r\n",
      "8\t24.806829415519122\t23.84708362897539\t27.371445191870418\t23.97464176363507\r\n",
      "9\t25.44816858143493\t23.66129679773161\t26.551531006998008\t24.33900361383545\r\n",
      "10-11\t25.180336963636517\t23.81916683561684\t27.038851703293865\t23.961644497452774\r\n",
      "12-13\t25.28368241909123\t24.05615066104491\t26.671738572265003\t23.988428347598852\r\n",
      "14-15\t25.178066239172807\t23.657573792097228\t26.668214400772783\t24.496145567957186\r\n",
      "16-17\t25.00745952423801\t23.838191060734047\t26.82627815852618\t24.328071256501765\r\n",
      "18-19\t25.467193068387576\t23.851775946944\t26.502917905944713\t24.178113078723708\r\n",
      "20-21\t25.289561069437188\t23.691718817124023\t26.355086677770206\t24.663633435668583\r\n",
      "22-23\t25.103784088004904\t23.85637376741577\t26.633098711675178\t24.406743432904147\r\n",
      "24-25\t25.49489947830214\t23.804703564605344\t26.355711283323824\t24.344685673768694\r\n",
      "26-27\t25.661185399614777\t23.70892797734052\t26.096147437045104\t24.533739185999607\r\n",
      "28-29\t25.64145564288766\t23.68327520343638\t26.210302875517073\t24.46496627815889\r\n",
      "30-31\t25.828676749236994\t23.619960914042316\t25.96418123184212\t24.58718110487857\r\n",
      "32-33\t25.678514366792943\t23.545420402461254\t25.965719040276408\t24.810346190469392\r\n",
      "34-35\t25.594759444704774\t23.71573428798476\t26.23971218575429\t24.449794081556174\r\n",
      "36-37\t25.747388085479155\t23.619416819106352\t25.974132057237664\t24.659063038176825\r\n",
      "38-39\t25.90799666403502\t23.38192748029126\t25.863866958638347\t24.846208897035368\r\n",
      "40-41\t25.62771962289016\t23.58223597640036\t26.09199179550678\t24.6980526052027\r\n",
      "42-43\t26.015650940308465\t23.570242739190732\t25.774343630406715\t24.639762690094084\r\n",
      "44-45\t25.866928374075428\t23.32603384624525\t25.638252373343732\t25.168785406335587\r\n",
      "46-47\t25.660682064873992\t23.394187495070945\t26.15305827495608\t24.792072165098983\r\n",
      "48-49\t25.6688520796397\t23.41429293202767\t26.034902489051188\t24.881952499281443\r\n",
      "50-51\t25.813414382724265\t23.20374305604632\t25.89194084110291\t25.090901720126507\r\n",
      "52-53\t25.81412648496217\t23.351133542413795\t26.06086016641493\t24.773879806209113\r\n",
      "54-55\t25.911507538619283\t23.358658112385758\t25.85545049078349\t24.87438385821147\r\n",
      "56-57\t25.93607603047338\t23.238912412545456\t25.83454595687392\t24.99046560010724\r\n",
      "58-59\t25.794997862087627\t23.54891452793012\t25.897308535663182\t24.758779074319076\r\n",
      "60-61\t25.957755944817062\t23.518565211001256\t25.633772255845965\t24.889906588335716\r\n",
      "62-63\t25.93386292558943\t23.602787737063878\t25.461443459983197\t25.001905877363495\r\n",
      "64-65\t25.8991154129563\t23.69641814706537\t25.612032937398137\t24.79243350258019\r\n",
      "66-67\t26.055772267721334\t23.66481827777764\t25.460020544266083\t24.81938891023494\r\n",
      "68-69\t26.081435592959025\t23.399733711669228\t25.47018585509\t25.048644840281746\r\n",
      "70-71\t26.092376496391072\t23.36516760937198\t25.7381916457826\t24.80426424845435\r\n",
      "72-73\t26.199481280517816\t23.377843142614317\t25.534809217556116\t24.887866359311754\r\n",
      "74-75\t26.066693878644582\t23.430658300266582\t25.37202587147075\t25.13062194961809\r\n",
      "76-77\t25.868333583776344\t23.75179896042242\t25.617583547626943\t24.762283908174297\r\n",
      "78-79\t26.020395593061806\t23.75367514197389\t25.333591664120792\t24.892337600843504\r\n",
      "80-81\t26.13111676505755\t23.60907307933699\t25.200417130236417\t25.059393025369044\r\n",
      "82-83\t26.024090747986676\t23.719312670821974\t25.407233421412435\t24.849363159778914\r\n",
      "84-85\t26.237988671694453\t23.8099675651798\t25.0522301850577\t24.89981357806805\r\n",
      "86-87\t26.251159460825253\t23.736218105199143\t24.95301986391496\t25.059602570060648\r\n",
      "88-89\t26.074964885003904\t23.949814437033147\t25.186316046710296\t24.78890463125266\r\n",
      "90-91\t26.287601383842187\t24.038957901666656\t24.86663815217192\t24.806802562319238\r\n",
      "92-93\t26.258363262780914\t24.047511674128028\t24.813753010292515\t24.880372052798542\r\n",
      "94-95\t26.16065508267739\t24.33975773069378\t24.713274012470805\t24.786313174158025\r\n",
      "96-97\t26.344909584045258\t24.34214061704753\t24.539307697770766\t24.773642101136446\r\n",
      "98-99\t26.241509520456457\t24.3301979070153\t24.51130710083126\t24.916985471696982\r\n",
      "100\t26.15715168505676\t24.721137678100032\t24.669686801058262\t24.45202383578494\r\n",
      ">>END_MODULE\r\n",
      ">>Per sequence GC content\twarn\r\n",
      "#GC Content\tCount\r\n",
      "0\t134790.0\r\n",
      "1\t88758.5\r\n",
      "2\t38474.0\r\n",
      "3\t31684.5\r\n",
      "4\t27710.0\r\n",
      "5\t25249.5\r\n",
      "6\t23465.5\r\n",
      "7\t22160.5\r\n",
      "8\t21230.0\r\n",
      "9\t21020.5\r\n",
      "10\t21054.5\r\n",
      "11\t20929.0\r\n",
      "12\t20831.5\r\n",
      "13\t20791.5\r\n",
      "14\t21015.0\r\n",
      "15\t21638.5\r\n",
      "16\t22632.0\r\n",
      "17\t23870.5\r\n",
      "18\t26081.0\r\n",
      "19\t29897.0\r\n",
      "20\t35794.0\r\n",
      "21\t43187.5\r\n",
      "22\t52976.0\r\n",
      "23\t65681.5\r\n",
      "24\t80252.5\r\n",
      "25\t98384.5\r\n",
      "26\t124038.0\r\n",
      "27\t154081.0\r\n",
      "28\t187601.5\r\n",
      "29\t226785.0\r\n",
      "30\t267110.5\r\n",
      "31\t315228.0\r\n",
      "32\t385294.5\r\n",
      "33\t473369.5\r\n",
      "34\t565896.5\r\n",
      "35\t683979.5\r\n",
      "36\t816084.5\r\n",
      "37\t969245.5\r\n",
      "38\t1166056.0\r\n",
      "39\t1389664.0\r\n",
      "40\t1617077.0\r\n",
      "41\t1823828.0\r\n",
      "42\t1987147.5\r\n",
      "43\t2153172.0\r\n",
      "44\t2304198.0\r\n",
      "45\t2427509.0\r\n",
      "46\t2530687.0\r\n",
      "47\t2555642.5\r\n",
      "48\t2535212.5\r\n",
      "49\t2561586.0\r\n",
      "50\t2594413.0\r\n",
      "51\t2554348.5\r\n",
      "52\t2469543.0\r\n",
      "53\t2341984.0\r\n",
      "54\t2244143.5\r\n",
      "55\t2186963.5\r\n",
      "56\t2076753.0\r\n",
      "57\t1953009.0\r\n",
      "58\t1886057.5\r\n",
      "59\t1827925.5\r\n",
      "60\t1738746.0\r\n",
      "61\t1618354.0\r\n",
      "62\t1495305.5\r\n",
      "63\t1400275.0\r\n",
      "64\t1305070.0\r\n",
      "65\t1182706.0\r\n",
      "66\t1032941.0\r\n",
      "67\t891794.5\r\n",
      "68\t769732.5\r\n",
      "69\t647565.5\r\n",
      "70\t531428.0\r\n",
      "71\t429547.0\r\n",
      "72\t349954.0\r\n",
      "73\t290864.5\r\n",
      "74\t244100.0\r\n",
      "75\t203934.0\r\n",
      "76\t172815.5\r\n",
      "77\t146330.5\r\n",
      "78\t122748.5\r\n",
      "79\t100820.5\r\n",
      "80\t80880.0\r\n",
      "81\t65463.0\r\n",
      "82\t57266.5\r\n",
      "83\t51837.0\r\n",
      "84\t42658.0\r\n",
      "85\t33830.5\r\n",
      "86\t28878.5\r\n",
      "87\t24247.5\r\n",
      "88\t18556.5\r\n",
      "89\t15256.0\r\n",
      "90\t11618.5\r\n",
      "91\t7501.0\r\n",
      "92\t4682.5\r\n",
      "93\t2767.5\r\n",
      "94\t1532.0\r\n",
      "95\t971.5\r\n",
      "96\t715.5\r\n",
      "97\t365.5\r\n",
      "98\t235.5\r\n",
      "99\t168.5\r\n",
      "100\t295.0\r\n",
      ">>END_MODULE\r\n",
      ">>Per base N content\tpass\r\n",
      "#Base\tN-Count\r\n",
      "1\t4.382584427567275E-6\r\n",
      "2\t0.0655210980536066\r\n",
      "3\t1.0226030330990307E-5\r\n",
      "4\t4.382584427567275E-6\r\n",
      "5\t4.382584427567275E-6\r\n",
      "6\t1.1686891806846066E-5\r\n",
      "7\t4.382584427567275E-6\r\n",
      "8\t7.30430737927879E-5\r\n",
      "9\t4.382584427567275E-6\r\n",
      "10-11\t9.144992838857046E-4\r\n",
      "12-13\t0.0017099383574891646\r\n",
      "14-15\t4.382584427567275E-6\r\n",
      "16-17\t4.382584427567275E-6\r\n",
      "18-19\t0.002360752144982905\r\n",
      "20-21\t0.1339062150306284\r\n",
      "22-23\t0.030717534252819024\r\n",
      "24-25\t5.836141596043753E-4\r\n",
      "26-27\t2.1912923098188725E-6\r\n",
      "28-29\t6.712658775745146E-4\r\n",
      "30-31\t5.8434461595169934E-5\r\n",
      "32-33\t8.76516923927549E-6\r\n",
      "34-35\t0.002023293232732759\r\n",
      "36-37\t0.0\r\n",
      "38-39\t0.0\r\n",
      "40-41\t2.7756370474152143E-5\r\n",
      "42-43\t0.0012103238388334236\r\n",
      "44-45\t3.323460148878744E-4\r\n",
      "46-47\t3.191982604527497E-4\r\n",
      "48-49\t0.0020298671986228633\r\n",
      "50-51\t9.736642590011792E-4\r\n",
      "52-53\t0.0\r\n",
      "54-55\t2.4250302624785557E-4\r\n",
      "56-57\t9.086559176275069E-4\r\n",
      "58-59\t6.763789226069707E-4\r\n",
      "60-61\t7.925174201172389E-4\r\n",
      "62-63\t4.1634555711228215E-5\r\n",
      "64-65\t6.924484002499008E-4\r\n",
      "66-67\t1.1321677430246269E-4\r\n",
      "68-69\t0.0025762294384824897\r\n",
      "70-71\t9.422557345172702E-4\r\n",
      "72-73\t0.0010613159629875224\r\n",
      "74-75\t7.88865277631712E-5\r\n",
      "76-77\t0.0010503595085503718\r\n",
      "78-79\t6.793006607002399E-4\r\n",
      "80-81\t3.9443264457797103E-4\r\n",
      "82-83\t6.668833416660882E-4\r\n",
      "84-85\t3.4695464106395606E-4\r\n",
      "86-87\t0.0017756773314241627\r\n",
      "88-89\t7.538046096379003E-4\r\n",
      "90-91\t0.0014842354329304393\r\n",
      "92-93\t1.4900788795167797E-4\r\n",
      "94-95\t0.0016325128900588248\r\n",
      "96-97\t2.169379545178841E-4\r\n",
      "98-99\t1.833381366464273E-4\r\n",
      "100\t0.0015368264522075024\r\n",
      ">>END_MODULE\r\n",
      ">>Sequence Length Distribution\twarn\r\n",
      "#Length\tCount\r\n",
      "24-25\t3.0\r\n",
      "26-27\t0.0\r\n",
      "28-29\t0.0\r\n",
      "30-31\t0.0\r\n",
      "32-33\t0.0\r\n",
      "34-35\t0.0\r\n",
      "36-37\t0.0\r\n",
      "38-39\t3.0\r\n",
      "40-41\t0.0\r\n",
      "42-43\t0.0\r\n",
      "44-45\t0.0\r\n",
      "46-47\t0.0\r\n",
      "48-49\t0.0\r\n",
      "50-51\t0.0\r\n",
      "52-53\t0.0\r\n",
      "54-55\t0.0\r\n",
      "56-57\t0.0\r\n",
      "58-59\t0.0\r\n",
      "60-61\t0.0\r\n",
      "62-63\t0.0\r\n",
      "64-65\t0.0\r\n",
      "66-67\t0.0\r\n",
      "68-69\t0.0\r\n",
      "70-71\t0.0\r\n",
      "72-73\t1.0\r\n",
      "74-75\t0.0\r\n",
      "76-77\t0.0\r\n",
      "78-79\t1.0\r\n",
      "80-81\t0.0\r\n",
      "82-83\t0.0\r\n",
      "84-85\t0.0\r\n",
      "86-87\t0.0\r\n",
      "88-89\t0.0\r\n",
      "90-91\t0.0\r\n",
      "92-93\t0.0\r\n",
      "94-95\t0.0\r\n",
      "96-97\t0.0\r\n",
      "98-99\t0.0\r\n",
      "100-101\t6.8452752E7\r\n",
      ">>END_MODULE\r\n",
      ">>Sequence Duplication Levels\tfail\r\n",
      "#Total Deduplicated Percentage\t23.207905466360835\r\n",
      "#Duplication Level\tPercentage of deduplicated\tPercentage of total\r\n",
      "1\t54.7702727678844\t12.711033127638583\r\n",
      "2\t11.48388292490874\t5.330337386160747\r\n",
      "3\t8.8724313783694\t6.177316460579117\r\n",
      "4\t6.395308636387326\t5.936868730459122\r\n",
      "5\t4.469294186031323\t5.186147848538552\r\n",
      "6\t3.1222305394205434\t4.347625872183405\r\n",
      "7\t2.1862113902242966\t3.551617109166474\r\n",
      "8\t1.5466361665254584\t2.871534875486204\r\n",
      "9\t1.1040029663240503\t2.305943682932745\r\n",
      ">10\t5.465918761716602\t21.9620276992874\r\n",
      ">50\t0.3146809314166593\t5.0308925299019664\r\n",
      ">100\t0.2197197219113611\t10.370432398746471\r\n",
      ">500\t0.030227396932039487\t4.872220899954798\r\n",
      ">1k\t0.01859451137275397\t7.490496184089353\r\n",
      ">5k\t4.443740933811831E-4\t0.6592463667600725\r\n",
      ">10k+\t1.433464817358655E-4\t1.1962588281151172\r\n",
      ">>END_MODULE\r\n",
      ">>Overrepresented sequences\twarn\r\n",
      "#Sequence\tCount\tPercentage\tPossible Source\r\n",
      "TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\t220869\t0.32265901331078545\tNo Hit\r\n",
      "GCACACGTCTGAACTCCAGTCACTATAGCCTATATCTCGTATGCCGTCTT\t148018\t0.21623379393321762\tTruSeq Adapter, Index 9 (96% over 26bp)\r\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\t131391\t0.1919440501741639\tNo Hit\r\n",
      ">>END_MODULE\r\n",
      ">>Adapter Content\tpass\r\n",
      "#Position\tIllumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tSOLID Small RNA Adapter\r\n",
      "1\t5.405187460666305E-5\t4.382584427567274E-6\t0.0\t7.30430737927879E-6\t0.0\r\n",
      "2\t1.0518202626161458E-4\t4.382584427567274E-6\t0.0\t7.30430737927879E-6\t0.0\r\n",
      "3\t1.636164852958449E-4\t1.460861475855758E-5\t0.0\t7.30430737927879E-6\t0.0\r\n",
      "4\t2.3519869761277705E-4\t1.460861475855758E-5\t0.0\t1.7530337710269095E-5\t0.0\r\n",
      "5\t2.9071143369529585E-4\t4.52867057515285E-5\t0.0\t3.652153689639395E-5\t0.0\r\n",
      "6\t3.695979533915068E-4\t4.52867057515285E-5\t0.0\t3.652153689639395E-5\t4.382584427567274E-6\r\n",
      "7\t4.850060099841117E-4\t4.9669290179095774E-5\t0.0\t3.652153689639395E-5\t4.382584427567274E-6\r\n",
      "8\t5.785011444388802E-4\t4.9669290179095774E-5\t0.0\t5.989532051008608E-5\t4.382584427567274E-6\r\n",
      "9\t6.82222309224639E-4\t4.9669290179095774E-5\t0.0\t5.989532051008608E-5\t4.382584427567274E-6\r\n",
      "10-11\t7.968999350793161E-4\t5.405187460666305E-5\t0.0\t7.523436600657155E-5\t4.382584427567274E-6\r\n",
      "12-13\t9.941162343198432E-4\t5.551273608251881E-5\t0.0\t9.203427297891276E-5\t4.382584427567274E-6\r\n",
      "14-15\t0.0013250013586011725\t5.989532051008608E-5\t0.0\t1.1613848733053276E-4\t4.382584427567274E-6\r\n",
      "16-17\t0.0018881634575435673\t6.354747419972547E-5\t0.0\t1.4024270168215276E-4\t4.382584427567274E-6\r\n",
      "18-19\t0.0028172713561878293\t6.719962788936487E-5\t0.0\t1.460861475855758E-4\t4.382584427567274E-6\r\n",
      "20-21\t0.004275941539829803\t6.719962788936487E-5\t0.0\t1.504687320131431E-4\t4.382584427567274E-6\r\n",
      "22-23\t0.006545389842571724\t6.719962788936487E-5\t0.0\t1.5704260865449397E-4\t5.843445903423032E-6\r\n",
      "24-25\t0.009782658873068083\t7.523436600657155E-5\t0.0\t1.5777303939242186E-4\t5.843445903423032E-6\r\n",
      "26-27\t0.013918357711215735\t9.568642666855215E-5\t0.0\t1.701903619371958E-4\t5.843445903423032E-6\r\n",
      "28-29\t0.019166502563227546\t1.0518202626161457E-4\t0.0\t1.8333811521989763E-4\t5.843445903423032E-6\r\n",
      "30-31\t0.02537370297413866\t1.1321676437882125E-4\t0.0\t1.8845113038539277E-4\t5.843445903423032E-6\r\n",
      "32-33\t0.032611541156266015\t1.2928624061323457E-4\t0.0\t2.0452060661980612E-4\t5.843445903423032E-6\r\n",
      "34-35\t0.040858834618209695\t1.3001667135116246E-4\t0.0\t2.2132051359214733E-4\t5.843445903423032E-6\r\n",
      "36-37\t0.0500498445935562\t1.3439925577872974E-4\t0.0\t2.3446826687484917E-4\t5.843445903423032E-6\r\n",
      "38-39\t0.060213788311822634\t1.358601172545855E-4\t0.0\t2.432334357299837E-4\t6.573876641350911E-6\r\n",
      "40-41\t0.072252017303612\t1.504687320131431E-4\t0.0\t2.541898967989019E-4\t7.30430737927879E-6\r\n",
      "42-43\t0.086431138788268\t1.533904549648546E-4\t0.0\t2.680680808195316E-4\t7.30430737927879E-6\r\n",
      "44-45\t0.10313462890320274\t1.533904549648546E-4\t0.0\t2.71720234509171E-4\t1.0226030330990307E-5\r\n",
      "46-47\t0.12263639917513917\t1.6288605455791702E-4\t0.0\t2.936331566470074E-4\t1.3147753282701823E-5\r\n",
      "48-49\t0.14582757510434935\t1.7530337710269095E-4\t0.0\t3.3745900092268014E-4\t1.3147753282701823E-5\r\n",
      "50-51\t0.1737723942759941\t1.8187725374404188E-4\t0.0\t3.579110615846607E-4\t1.3147753282701823E-5\r\n",
      "52-53\t0.20817787332461102\t1.9429457628881581E-4\t0.0\t3.688675226535789E-4\t1.3147753282701823E-5\r\n",
      "54-55\t0.24984748606192067\t1.9429457628881581E-4\t0.0\t3.754413992949298E-4\t1.3147753282701823E-5\r\n",
      "56-57\t0.2989207447588673\t1.9721629924052733E-4\t0.0\t3.754413992949298E-4\t1.3147753282701823E-5\r\n",
      "58-59\t0.35832448538232786\t2.0963362178530128E-4\t0.0\t3.827457066742086E-4\t1.3147753282701823E-5\r\n",
      "60-61\t0.4290382155518638\t2.29355251709354E-4\t0.0\t4.0246733659826136E-4\t1.3147753282701823E-5\r\n",
      "62-63\t0.5149171487022584\t2.432334357299837E-4\t0.0\t4.199976743085304E-4\t1.3147753282701823E-5\r\n",
      "64-65\t0.6176076757168009\t2.512681738471904E-4\t0.0\t4.3168456611537647E-4\t1.3147753282701823E-5\r\n",
      "66-67\t0.7406969419494553\t2.512681738471904E-4\t0.0\t4.3971930423258317E-4\t1.3147753282701823E-5\r\n",
      "68-69\t0.886279092325861\t2.5638118901268557E-4\t0.0\t4.3971930423258317E-4\t1.3147753282701823E-5\r\n",
      "70-71\t1.057904750663085\t2.6587678860574794E-4\t0.0\t4.3971930423258317E-4\t1.3147753282701823E-5\r\n",
      "72-73\t1.2637306370115682\t2.673376500816037E-4\t0.0\t4.404497349705111E-4\t1.3147753282701823E-5\r\n",
      "74-75\t1.5069735683411452\t2.680680808195316E-4\t0.0\t4.4118016570843894E-4\t1.3147753282701823E-5\r\n",
      "76-77\t1.7959977070318276\t2.7025937303331524E-4\t0.0\t4.543279189911407E-4\t1.3147753282701823E-5\r\n",
      "78-79\t2.1312164768812827\t2.7025937303331524E-4\t0.0\t5.003450554805971E-4\t1.460861475855758E-5\r\n",
      "80-81\t2.518359376597817\t2.7025937303331524E-4\t0.0\t5.039972091702365E-4\t1.460861475855758E-5\r\n",
      "82-83\t2.965821246652436\t2.848679877918728E-4\t0.0\t5.098406550736595E-4\t1.460861475855758E-5\r\n",
      "84-85\t3.480216137377076\t2.877897107435843E-4\t0.0\t5.222579776184335E-4\t1.460861475855758E-5\r\n",
      "86-87\t4.060028404990536\t2.921722951711516E-4\t1.460861475855758E-6\t5.295622849977123E-4\t1.460861475855758E-5\r\n",
      "88-89\t4.709852458834384\t2.965548795987189E-4\t1.460861475855758E-6\t5.346753001632074E-4\t1.460861475855758E-5\r\n",
      ">>END_MODULE\r\n"
     ]
    }
   ],
   "source": [
    "cat SRR7191189_1_fastqc/fastqc_data.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_overseq(filename, withoutAT = True):\n",
    "    \n",
    "    f = open(filename,'r')\n",
    "    txtread = f.read()\n",
    "    sequence = []\n",
    "    in_code_section = False\n",
    "    # id filename.split('/')[2]\n",
    "    \n",
    "    for line in txtread.split('\\n'):\n",
    "        line = line.strip()\n",
    "        \n",
    "        if line == '#Sequence\tCount\tPercentage\tPossible Source':\n",
    "            in_code_section = True\n",
    "            continue\n",
    "        \n",
    "        if line.startswith(\">>END_MODULE\"):\n",
    "            in_code_section = in_code_section*-1\n",
    "            \n",
    "        if in_code_section and not line.startswith(\">>END_MODULE\"):\n",
    "            line.split('\\t')\n",
    "        \n",
    "            tmp_seq = line.split('\\t')[0]\n",
    "            \n",
    "            if (tmp_seq == 'T'*50 or tmp_seq == 'A'*50) and withoutAT:\n",
    "                continue\n",
    "            else:\n",
    "                sequence.append(tmp_seq)\n",
    "        \n",
    "        if in_code_section == -1:\n",
    "            return sequence\n",
    "\n",
    "    return sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "Basic_stat = None\n",
    "df = []\n",
    "\n",
    "\n",
    "for i in identify:\n",
    "\n",
    "    filename = './summary/' + i + '/summary.txt'\n",
    "    tmp = pd.read_csv(filename, delimiter='\\t')\n",
    "    new_row = [j for j in tmp.PASS]\n",
    "    new_row.insert(0,i[0:-7])\n",
    "    df.append(new_row)\n",
    "    \n",
    "    if Basic_stat is None:\n",
    "\n",
    "        Basic_stat = list(tmp.iloc[:,1])\n",
    "        Basic_stat.insert(0,'id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./summary/SRR7191194_2_fastqc/fastqc_data.txt'"
      ]
     },
     "execution_count": 502,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sequences = []\n",
    "\n",
    "for i in identify:\n",
    "\n",
    "    filename = './summary/' + i + '/fastqc_data.txt'\n",
    "    sequences.append(find_overseq(filename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 541,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[],\n",
       " [],\n",
       " [],\n",
       " ['GCACACGTCTGAACTCCAGTCACCCTATCCTATATCTCGTATGCCGTCTT'],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " ['GCACACGTCTGAACTCCAGTCACTATAGCCTATATCTCGTATGCCGTCTT'],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " [],\n",
       " []]"
      ]
     },
     "execution_count": 541,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(df,columns=Basic_stat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 542,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['sequences_withoutAT'] = sequences "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "metadata": {},
   "outputs": [],
   "source": [
    "# onlyfiles = [f for f in listdir('testout') if isfile(join('testout', f))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_adapter(filename):\n",
    "    \n",
    "    with open(filename,'r') as f:\n",
    "        txtread = f.read()\n",
    "    sequence = []\n",
    "    in_code_section = False\n",
    "    # id filename.split('/')[2]\n",
    "    \n",
    "    for line in txtread.split('\\n'):\n",
    "        line = line.strip()\n",
    "        \n",
    "        if line == \"#Position\tIllumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tSOLID Small RNA Adapter\":\n",
    "            in_code_section = True\n",
    "            continue\n",
    "        \n",
    "        if line.startswith(\">>END_MODULE\"):\n",
    "            in_code_section = in_code_section*-1\n",
    "            \n",
    "        if in_code_section and not line.startswith(\">>END_MODULE\"):\n",
    "            line.split('\\t')\n",
    "            sequence.append(line.split('\\t'))\n",
    "        \n",
    "        if in_code_section == -1:\n",
    "            return sequence[-1][1:6]\n",
    "\n",
    "    return 'Cannot find anything'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "metadata": {},
   "outputs": [],
   "source": [
    "adapters = []\n",
    "\n",
    "for i in identify:\n",
    "\n",
    "    filename = './summary/' + i + '/fastqc_data.txt'\n",
    "    adapter = find_adapter(filename)\n",
    "    adapters.append([float(i)>1 for i in adapter])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 569,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['adapters'] = adapters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 570,
   "metadata": {},
   "outputs": [],
   "source": [
    "adapter_names = \"Illumina Universal Adapter\tIllumina Small RNA 3' Adapter\tIllumina Small RNA 5' Adapter\tNextera Transposase Sequence\tSOLID Small RNA Adapter\".split('\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 571,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Illumina Universal Adapter',\n",
       " \"Illumina Small RNA 3' Adapter\",\n",
       " \"Illumina Small RNA 5' Adapter\",\n",
       " 'Nextera Transposase Sequence',\n",
       " 'SOLID Small RNA Adapter']"
      ]
     },
     "execution_count": 571,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adapter_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set adapter dictionary (can be found in fastqc adapter_list.txt) could be generated automatically\n",
    "dictionary = {'Illumina Universal Adapter':'AGATCGGAAGAG',\n",
    "              \"Illumina Small RNA 3' Adapter\":'TGGAATTCTCGG',\n",
    "              \"Illumina Small RNA 5' Adapter\":'GATCGTCGGACT',\n",
    "              'Nextera Transposase Sequence':'CTGTCTCTTATA',\n",
    "              'SOLID Small RNA Adapter':'CGCCTTGGCCGT',\n",
    "             }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 600,
   "metadata": {},
   "outputs": [],
   "source": [
    "adapter_seqs = []\n",
    "\n",
    "for adapter_list in adapters:\n",
    "   \n",
    "    adapter_seq = []\n",
    "    \n",
    "    for i,adapter_name in zip(adapter_list,adapter_names):\n",
    "        \n",
    "        if i == True:\n",
    "            adapter_seq.append(dictionary[adapter_name])\n",
    "            \n",
    "    adapter_seqs.append(adapter_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['adapter_seqs'] = adapter_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run cutadapter to cut off the low quality data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt version 2.8\r\n",
      "\r\n",
      "Copyright (C) 2010-2019 Marcel Martin <marcel.martin@scilifelab.se>\r\n",
      "\r\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\r\n",
      "\r\n",
      "Usage:\r\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\r\n",
      "\r\n",
      "For paired-end reads:\r\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\r\n",
      "\r\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\r\n",
      "characters are supported. All reads from input.fastq will be written to\r\n",
      "output.fastq with the adapter sequence removed. Adapter matching is\r\n",
      "error-tolerant. Multiple adapter sequences can be given (use further -a\r\n",
      "options), but only the best-matching adapter will be removed.\r\n",
      "\r\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\r\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\r\n",
      "standard input/output. Without the -o option, output is sent to standard output.\r\n",
      "\r\n",
      "Citation:\r\n",
      "\r\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\r\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\r\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\r\n",
      "\r\n",
      "Run \"cutadapt --help\" to see all command-line options.\r\n",
      "See https://cutadapt.readthedocs.io/ for full documentation.\r\n",
      "\r\n",
      "Options:\r\n",
      "  -h, --help            Show this help message and exit\r\n",
      "  --version             Show version number and exit\r\n",
      "  --debug               Print debugging information.\r\n",
      "  -j CORES, --cores CORES\r\n",
      "                        Number of CPU cores to use. Use 0 to auto-detect.\r\n",
      "                        Default: 1\r\n",
      "\r\n",
      "Finding adapters:\r\n",
      "  Parameters -a, -g, -b specify adapters to be removed from each read (or\r\n",
      "  from the first read in a pair if data is paired). If specified multiple\r\n",
      "  times, only the best matching adapter is trimmed (but see the --times\r\n",
      "  option). When the special notation 'file:FILE' is used, adapter sequences\r\n",
      "  are read from the given FASTA file.\r\n",
      "\r\n",
      "  -a ADAPTER, --adapter ADAPTER\r\n",
      "                        Sequence of an adapter ligated to the 3' end (paired\r\n",
      "                        data: of the first read). The adapter and subsequent\r\n",
      "                        bases are trimmed. If a '$' character is appended\r\n",
      "                        ('anchoring'), the adapter is only found if it is a\r\n",
      "                        suffix of the read.\r\n",
      "  -g ADAPTER, --front ADAPTER\r\n",
      "                        Sequence of an adapter ligated to the 5' end (paired\r\n",
      "                        data: of the first read). The adapter and any\r\n",
      "                        preceding bases are trimmed. Partial matches at the 5'\r\n",
      "                        end are allowed. If a '^' character is prepended\r\n",
      "                        ('anchoring'), the adapter is only found if it is a\r\n",
      "                        prefix of the read.\r\n",
      "  -b ADAPTER, --anywhere ADAPTER\r\n",
      "                        Sequence of an adapter that may be ligated to the 5'\r\n",
      "                        or 3' end (paired data: of the first read). Both types\r\n",
      "                        of matches as described under -a und -g are allowed.\r\n",
      "                        If the first base of the read is part of the match,\r\n",
      "                        the behavior is as with -g, otherwise as with -a. This\r\n",
      "                        option is mostly for rescuing failed library\r\n",
      "                        preparations - do not use if you know which end your\r\n",
      "                        adapter was ligated to!\r\n",
      "  -e RATE, --error-rate RATE\r\n",
      "                        Maximum allowed error rate as value between 0 and 1\r\n",
      "                        (no. of errors divided by length of matching region).\r\n",
      "                        Default: 0.1 (=10%)\r\n",
      "  --no-indels           Allow only mismatches in alignments. Default: allow\r\n",
      "                        both mismatches and indels\r\n",
      "  -n COUNT, --times COUNT\r\n",
      "                        Remove up to COUNT adapters from each read. Default: 1\r\n",
      "  -O MINLENGTH, --overlap MINLENGTH\r\n",
      "                        Require MINLENGTH overlap between read and adapter for\r\n",
      "                        an adapter to be found. Default: 3\r\n",
      "  --match-read-wildcards\r\n",
      "                        Interpret IUPAC wildcards in reads. Default: False\r\n",
      "  -N, --no-match-adapter-wildcards\r\n",
      "                        Do not interpret IUPAC wildcards in adapters.\r\n",
      "  --action {trim,mask,lowercase,none}\r\n",
      "                        What to do with found adapters. mask: replace with 'N'\r\n",
      "                        characters; lowercase: convert to lowercase; none:\r\n",
      "                        leave unchanged (useful with --discard-untrimmed).\r\n",
      "                        Default: trim\r\n",
      "  --rc, --revcomp       Check both the read and its reverse complement for\r\n",
      "                        adapter matches. If match is on reverse-complemented\r\n",
      "                        version, output that one. Default: check only read\r\n",
      "\r\n",
      "Additional read modifications:\r\n",
      "  -u LENGTH, --cut LENGTH\r\n",
      "                        Remove bases from each read (first read only if\r\n",
      "                        paired). If LENGTH is positive, remove bases from the\r\n",
      "                        beginning. If LENGTH is negative, remove bases from\r\n",
      "                        the end. Can be used twice if LENGTHs have different\r\n",
      "                        signs. This is applied *before* adapter trimming.\r\n",
      "  --nextseq-trim 3'CUTOFF\r\n",
      "                        NextSeq-specific quality trimming (each read). Trims\r\n",
      "                        also dark cycles appearing as high-quality G bases.\r\n",
      "  -q [5'CUTOFF,]3'CUTOFF, --quality-cutoff [5'CUTOFF,]3'CUTOFF\r\n",
      "                        Trim low-quality bases from 5' and/or 3' ends of each\r\n",
      "                        read before adapter removal. Applied to both reads if\r\n",
      "                        data is paired. If one value is given, only the 3' end\r\n",
      "                        is trimmed. If two comma-separated cutoffs are given,\r\n",
      "                        the 5' end is trimmed with the first cutoff, the 3'\r\n",
      "                        end with the second.\r\n",
      "  --quality-base N      Assume that quality values in FASTQ are encoded as\r\n",
      "                        ascii(quality + N). This needs to be set to 64 for\r\n",
      "                        some old Illumina FASTQ files. Default: 33\r\n",
      "  --length LENGTH, -l LENGTH\r\n",
      "                        Shorten reads to LENGTH. Positive values remove bases\r\n",
      "                        at the end while negative ones remove bases at the\r\n",
      "                        beginning. This and the following modifications are\r\n",
      "                        applied after adapter trimming.\r\n",
      "  --trim-n              Trim N's on ends of reads.\r\n",
      "  --length-tag TAG      Search for TAG followed by a decimal number in the\r\n",
      "                        description field of the read. Replace the decimal\r\n",
      "                        number with the correct length of the trimmed read.\r\n",
      "                        For example, use --length-tag 'length=' to correct\r\n",
      "                        fields like 'length=123'.\r\n",
      "  --strip-suffix STRIP_SUFFIX\r\n",
      "                        Remove this suffix from read names if present. Can be\r\n",
      "                        given multiple times.\r\n",
      "  -x PREFIX, --prefix PREFIX\r\n",
      "                        Add this prefix to read names. Use {name} to insert\r\n",
      "                        the name of the matching adapter.\r\n",
      "  -y SUFFIX, --suffix SUFFIX\r\n",
      "                        Add this suffix to read names; can also include {name}\r\n",
      "  --zero-cap, -z        Change negative quality values to zero.\r\n",
      "\r\n",
      "Filtering of processed reads:\r\n",
      "  Filters are applied after above read modifications. Paired-end reads are\r\n",
      "  always discarded pairwise (see also --pair-filter).\r\n",
      "\r\n",
      "  -m LEN[:LEN2], --minimum-length LEN[:LEN2]\r\n",
      "                        Discard reads shorter than LEN. Default: 0\r\n",
      "  -M LEN[:LEN2], --maximum-length LEN[:LEN2]\r\n",
      "                        Discard reads longer than LEN. Default: no limit\r\n",
      "  --max-n COUNT         Discard reads with more than COUNT 'N' bases. If COUNT\r\n",
      "                        is a number between 0 and 1, it is interpreted as a\r\n",
      "                        fraction of the read length.\r\n",
      "  --discard-trimmed, --discard\r\n",
      "                        Discard reads that contain an adapter. Use also -O to\r\n",
      "                        avoid discarding too many randomly matching reads.\r\n",
      "  --discard-untrimmed, --trimmed-only\r\n",
      "                        Discard reads that do not contain an adapter.\r\n",
      "  --discard-casava      Discard reads that did not pass CASAVA filtering\r\n",
      "                        (header has :Y:).\r\n",
      "\r\n",
      "Output:\r\n",
      "  --quiet               Print only error messages.\r\n",
      "  --report {full,minimal}\r\n",
      "                        Which type of report to print: 'full' or 'minimal'.\r\n",
      "                        Default: full\r\n",
      "  -o FILE, --output FILE\r\n",
      "                        Write trimmed reads to FILE. FASTQ or FASTA format is\r\n",
      "                        chosen depending on input. Summary report is sent to\r\n",
      "                        standard output. Use '{name}' for demultiplexing (see\r\n",
      "                        docs). Default: write to standard output\r\n",
      "  --fasta               Output FASTA to standard output even on FASTQ input.\r\n",
      "  -Z                    Use compression level 1 for gzipped output files\r\n",
      "                        (faster, but uses more space)\r\n",
      "  --info-file FILE      Write information about each read and its adapter\r\n",
      "                        matches into FILE. See the documentation for the file\r\n",
      "                        format.\r\n",
      "  -r FILE, --rest-file FILE\r\n",
      "                        When the adapter matches in the middle of a read,\r\n",
      "                        write the rest (after the adapter) to FILE.\r\n",
      "  --wildcard-file FILE  When the adapter has N wildcard bases, write adapter\r\n",
      "                        bases matching wildcard positions to FILE. (Inaccurate\r\n",
      "                        with indels.)\r\n",
      "  --too-short-output FILE\r\n",
      "                        Write reads that are too short (according to length\r\n",
      "                        specified by -m) to FILE. Default: discard reads\r\n",
      "  --too-long-output FILE\r\n",
      "                        Write reads that are too long (according to length\r\n",
      "                        specified by -M) to FILE. Default: discard reads\r\n",
      "  --untrimmed-output FILE\r\n",
      "                        Write reads that do not contain any adapter to FILE.\r\n",
      "                        Default: output to same file as trimmed reads\r\n",
      "\r\n",
      "Paired-end options:\r\n",
      "  The -A/-G/-B/-U options work like their -a/-b/-g/-u counterparts, but are\r\n",
      "  applied to the second read in each pair.\r\n",
      "\r\n",
      "  -A ADAPTER            3' adapter to be removed from second read in a pair.\r\n",
      "  -G ADAPTER            5' adapter to be removed from second read in a pair.\r\n",
      "  -B ADAPTER            5'/3 adapter to be removed from second read in a pair.\r\n",
      "  -U LENGTH             Remove LENGTH bases from second read in a pair.\r\n",
      "  -p FILE, --paired-output FILE\r\n",
      "                        Write second read in a pair to FILE.\r\n",
      "  --pair-adapters       Treat adapters given with -a/-A etc. as pairs. Either\r\n",
      "                        both or none are removed from each read pair.\r\n",
      "  --pair-filter (any|both|first)\r\n",
      "                        Which of the reads in a paired-end read have to match\r\n",
      "                        the filtering criterion in order for the pair to be\r\n",
      "                        filtered. Default: any\r\n",
      "  --interleaved         Read and write interleaved paired-end reads.\r\n",
      "  --untrimmed-paired-output FILE\r\n",
      "                        Write second read in a pair to this FILE when no\r\n",
      "                        adapter was found. Use with --untrimmed-output.\r\n",
      "                        Default: output to same file as trimmed reads\r\n",
      "  --too-short-paired-output FILE\r\n",
      "                        Write second read in a pair to this file if pair is\r\n",
      "                        too short. Use also --too-short-output.\r\n",
      "  --too-long-paired-output FILE\r\n",
      "                        Write second read in a pair to this file if pair is\r\n",
      "                        too long. Use also --too-long-output.\r\n"
     ]
    }
   ],
   "source": [
    "!cutadapt --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO: test cutadapter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir cutpair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Per base sequence quality</th>\n",
       "      <th>Per sequence quality scores</th>\n",
       "      <th>Per base sequence content</th>\n",
       "      <th>Per sequence GC content</th>\n",
       "      <th>Per base N content</th>\n",
       "      <th>Sequence Length Distribution</th>\n",
       "      <th>Sequence Duplication Levels</th>\n",
       "      <th>Overrepresented sequences</th>\n",
       "      <th>Adapter Content</th>\n",
       "      <th>sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR7191194_2</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>PASS</td>\n",
       "      <td>WARN</td>\n",
       "      <td>FAIL</td>\n",
       "      <td>WARN</td>\n",
       "      <td>PASS</td>\n",
       "      <td>[TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id Per base sequence quality Per sequence quality scores  \\\n",
       "15  SRR7191194_2                      PASS                        PASS   \n",
       "\n",
       "   Per base sequence content Per sequence GC content Per base N content  \\\n",
       "15                      PASS                    PASS               PASS   \n",
       "\n",
       "   Sequence Length Distribution Sequence Duplication Levels  \\\n",
       "15                         WARN                        FAIL   \n",
       "\n",
       "   Overrepresented sequences Adapter Content  \\\n",
       "15                      WARN            PASS   \n",
       "\n",
       "                                            sequences  \n",
       "15  [TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...  "
      ]
     },
     "execution_count": 384,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input1 = i + '_1.fastq'\n",
    "input2 = i + '_2.fastq'\n",
    "\n",
    "df.loc[df.id == i[0:12]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parameter_generation(sequence1, parameter = '-a'):\n",
    "    adapters1 = []\n",
    "    if len(sequence1.iloc[0]) is not 0:\n",
    "        for j in sequence1.iloc[0]:\n",
    "            adapters1.append(parameter)\n",
    "            adapters1.append(j)\n",
    "    return adapters1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt -q 20 -a GCACACGTCTGAACTCCAGTCACTATAGCCTATATCTCGTATGCCGTCTT -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191189_1.fastq -p cutpair/SRR7191189_2.fastq -j 80 SRR7191189_1.fastq SRR7191189_2.fastq\n",
      "cutadapt -q 20 -a GCACACGTCTGAACTCCAGTCACCCTATCCTATATCTCGTATGCCGTCTT -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191190_1.fastq -p cutpair/SRR7191190_2.fastq -j 80 SRR7191190_1.fastq SRR7191190_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191191_1.fastq -p cutpair/SRR7191191_2.fastq -j 80 SRR7191191_1.fastq SRR7191191_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191192_1.fastq -p cutpair/SRR7191192_2.fastq -j 80 SRR7191192_1.fastq SRR7191192_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191193_1.fastq -p cutpair/SRR7191193_2.fastq -j 80 SRR7191193_1.fastq SRR7191193_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191194_1.fastq -p cutpair/SRR7191194_2.fastq -j 80 SRR7191194_1.fastq SRR7191194_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191195_1.fastq -p cutpair/SRR7191195_2.fastq -j 80 SRR7191195_1.fastq SRR7191195_2.fastq\n",
      "cutadapt -q 20 -a AGATCGGAAGAG -A AGATCGGAAGAG -o cutpair/SRR7191196_1.fastq -p cutpair/SRR7191196_2.fastq -j 80 SRR7191196_1.fastq SRR7191196_2.fastq\n"
     ]
    }
   ],
   "source": [
    "tmpdir = 'cutpair/'\n",
    "\n",
    "for i in data.Run:\n",
    "    \n",
    "    input1 = i + '_1'\n",
    "    input2 = i + '_2'   \n",
    "    input_dir1 = i + '_1.fastq'\n",
    "    input_dir2 = i + '_2.fastq'\n",
    "    output_dir1 = tmpdir + i + '_1.fastq'\n",
    "    output_dir2 = tmpdir + i + '_2.fastq'\n",
    "    \n",
    "    sequence1 = df.loc[df.id == input1].sequences_withoutAT # change here for AAA\n",
    "    sequence2 = df.loc[df.id == input2].sequences_withoutAT\n",
    "    \n",
    "    adapter_seq1 = df.loc[df.id == input1].adapter_seqs\n",
    "    adapter_seq2 = df.loc[df.id == input1].adapter_seqs\n",
    "    \n",
    "    # -a and -A for input1 and input2\n",
    "    over_seq1 = parameter_generation(sequence1)\n",
    "    over_seq2 = parameter_generation(sequence2,'-A')\n",
    "\n",
    "    # -a and -A for adapters\n",
    "    adapter1 = parameter_generation(adapter_seq1)\n",
    "    adapter2 = parameter_generation(adapter_seq2,'-A')\n",
    "    \n",
    "    # command generation\n",
    "    command = ['cutadapt', '-q', '20'] # cut down the low quality reads\n",
    "    command.extend(over_seq1)\n",
    "    command.extend(over_seq2)\n",
    "    command.extend(adapter1)\n",
    "    command.extend(adapter2)\n",
    "\n",
    "    command.extend(['-o', output_dir1, # output_dir for input1\n",
    "                    '-p', output_dir2, # output_dir for input2\n",
    "                    '-j', '80',\n",
    "                    input_dir1,\n",
    "                    input_dir2])\n",
    "    print(\" \".join(command))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cutadapt',\n",
       " '-q',\n",
       " '20',\n",
       " '-a',\n",
       " 'GCACACGTCTGAACTCCAGTCACTATAGCCTATATCTCGTATGCCGTCTT',\n",
       " '-a',\n",
       " True,\n",
       " '-a',\n",
       " False,\n",
       " '-a',\n",
       " False,\n",
       " '-a',\n",
       " False,\n",
       " '-a',\n",
       " False,\n",
       " '-A',\n",
       " True,\n",
       " '-A',\n",
       " False,\n",
       " '-A',\n",
       " False,\n",
       " '-A',\n",
       " False,\n",
       " '-A',\n",
       " False,\n",
       " '-o',\n",
       " 'cutpair/SRR7191189_1.fastq',\n",
       " '-p',\n",
       " 'cutpair/SRR7191189_2.fastq',\n",
       " '-j',\n",
       " '80',\n",
       " 'SRR7191189_1.fastq',\n",
       " 'SRR7191189_2.fastq']"
      ]
     },
     "execution_count": 606,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 586,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt -q 20 -a GCACACGTCTGAACTCCAGTCACTATAGCCTATATCTCGTATGCCGTCTT -o cutpair/SRR7191189_1.fastq -p cutpair/SRR7191189_2.fastq -j 80 SRR7191189_1.fastq SRR7191189_2.fastq\n",
      "cutadapt -q 20 -a GCACACGTCTGAACTCCAGTCACCCTATCCTATATCTCGTATGCCGTCTT -o cutpair/SRR7191190_1.fastq -p cutpair/SRR7191190_2.fastq -j 80 SRR7191190_1.fastq SRR7191190_2.fastq\n",
      "cutadapt -q 20 -o cutpair/SRR7191191_1.fastq -p cutpair/SRR7191191_2.fastq -j 80 SRR7191191_1.fastq SRR7191191_2.fastq\n",
      "cutadapt -q 20 -o cutpair/SRR7191192_1.fastq -p cutpair/SRR7191192_2.fastq -j 80 SRR7191192_1.fastq SRR7191192_2.fastq\n",
      "cutadapt -q 20 -o cutpair/SRR7191193_1.fastq -p cutpair/SRR7191193_2.fastq -j 80 SRR7191193_1.fastq SRR7191193_2.fastq\n",
      "cutadapt -q 20 -o cutpair/SRR7191194_1.fastq -p cutpair/SRR7191194_2.fastq -j 80 SRR7191194_1.fastq SRR7191194_2.fastq\n",
      "cutadapt -q 20 -o cutpair/SRR7191195_1.fastq -p cutpair/SRR7191195_2.fastq -j 80 SRR7191195_1.fastq SRR7191195_2.fastq\n",
      "cutadapt -q 20 -o cutpair/SRR7191196_1.fastq -p cutpair/SRR7191196_2.fastq -j 80 SRR7191196_1.fastq SRR7191196_2.fastq\n"
     ]
    }
   ],
   "source": [
    "tmpdir = 'cutpair/'\n",
    "\n",
    "for i in data.Run:\n",
    "    \n",
    "    input1 = i + '_1'\n",
    "    input2 = i + '_2'   \n",
    "    input_dir1 = i + '_1.fastq'\n",
    "    input_dir2 = i + '_2.fastq'\n",
    "    output_dir1 = tmpdir + i + '_1.fastq'\n",
    "    output_dir2 = tmpdir + i + '_2.fastq'\n",
    "    \n",
    "    sequence1 = df.loc[df.id == input1].sequences_withoutAT # change here for AAA\n",
    "    sequence2 = df.loc[df.id == input2].sequences_withoutAT\n",
    "\n",
    "    adapter_seq1 = df.loc[df.id == input1].adapters\n",
    "    adapter_seq2 = df.loc[df.id == input1].adapters\n",
    "    \n",
    "    adapters1 = []\n",
    "    if len(sequence1.iloc[0]) is not 0:\n",
    "        for j in sequence1.iloc[0]:\n",
    "            adapters1.append('-a')\n",
    "            adapters1.append(j)\n",
    "\n",
    "    adapters2 = []\n",
    "    if len(sequence2.iloc[0]) is not 0:\n",
    "        for j in sequence2.iloc[0]:\n",
    "            adapters2.append('-A')\n",
    "            adapters2.append(j)\n",
    "    \n",
    "    \n",
    "    command = ['cutadapt', '-q', '20'] # cut down the low quality reads\n",
    "    command.extend(adapters1)\n",
    "    command.extend(adapters2)\n",
    "\n",
    "    command.extend(['-o', output_dir1, # output_dir for input1\n",
    "                    '-p', output_dir2, # output_dir for input2\n",
    "                    '-j', '80',\n",
    "                    input_dir1,\n",
    "                    input_dir2])\n",
    "    print(\" \".join(command))\n",
    "\n",
    "#     subprocess.run(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 2.8 with Python 3.7.6\n",
      "Command line parameters: -q 20 -o cutpair/out.1.fastq -p cutpair/out.2.fastq SRR7191192_1.fastq SRR7191192_2.fastq\n",
      "Processing reads on 1 core in paired-end mode ...\n",
      "[------->8   ] 00:12:10    77,263,058 reads  @      9.5 s/read;   6.34 M reads/minute\n",
      "Finished in 731.00 s (9 us/read; 6.34 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total read pairs processed:         77,263,058\n",
      "  Read 1 with adapter:                       0 (0.0%)\n",
      "  Read 2 with adapter:                       0 (0.0%)\n",
      "Pairs written (passing filters):    77,263,058 (100.0%)\n",
      "\n",
      "Total basepairs processed: 15,452,610,605 bp\n",
      "  Read 1: 7,726,305,531 bp\n",
      "  Read 2: 7,726,305,074 bp\n",
      "Quality-trimmed:             201,314,526 bp (1.3%)\n",
      "  Read 1:    41,944,144 bp\n",
      "  Read 2:   159,370,382 bp\n",
      "Total written (filtered):  15,251,296,079 bp (98.7%)\n",
      "  Read 1: 7,684,361,387 bp\n",
      "  Read 2: 7,566,934,692 bp\n"
     ]
    }
   ],
   "source": [
    "!cutadapt -q 20 -o cutpair/out.1.fastq -p cutpair/out.2.fastq SRR7191192_1.fastq SRR7191192_2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 2.8 with Python 3.7.6\n",
      "Command line parameters: -q 20 -o cut/SRR7191192_2.fastq SRR7191192_2.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[---->8      ] 00:05:06    77,263,058 reads  @      4.0 s/read;  15.12 M reads/minute\n",
      "Finished in 306.68 s (4 us/read; 15.12 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              77,263,058\n",
      "Reads with adapters:                         0 (0.0%)\n",
      "Reads written (passing filters):    77,263,058 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,726,305,074 bp\n",
      "Quality-trimmed:             159,370,382 bp (2.1%)\n",
      "Total written (filtered):  7,566,934,692 bp (97.9%)\n"
     ]
    }
   ],
   "source": [
    "!cutadapt -q 20 -o cut/SRR7191192_2.fastq SRR7191192_2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "args_1 = list(data.Run + '_1' + '.fastq')\n",
    "args_2 = list(data.Run + '_2' + '.fastq')\n",
    "args_1 = args_1[0:-1]\n",
    "args_2 = args_2[0:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR --runThreadN 10 --genomeDir ./genome --readFilesIn ./data/SRR7191196_1.fastq ./data/SRR7191196_2.fastq --outFileNamePrefix ./test_sample/ --outSAMtype BAM SortedByCoordinate --sjdbGTFfile gencode.v33.primary_assembly.annotation.gtf --sjdbOverhang 100 --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomedir = '/mnt/Tank/junfan/project/test1/genome'\n",
    "GTFdir = '/mnt/Tank/junfan/project/test1/gencode.v33.primary_assembly.annotation.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir STAROUTPUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SRR7191189_1.fastq',\n",
       " 'SRR7191190_1.fastq',\n",
       " 'SRR7191191_1.fastq',\n",
       " 'SRR7191192_1.fastq',\n",
       " 'SRR7191193_1.fastq',\n",
       " 'SRR7191194_1.fastq',\n",
       " 'SRR7191195_1.fastq']"
      ]
     },
     "execution_count": 498,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "args_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "template = 'STAR --runThreadN 80 --genomeDir ' + genomedir + ' --readFilesIn {} {} --outFileNamePrefix ./{} --outSAMtype BAM SortedByCoordinate --sjdbGTFfile '+ GTFdir + ' --sjdbOverhang 100 --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded'\n",
    "\n",
    "parent_dir = \"/mnt/Tank/junfan/project/test\"\n",
    "\n",
    "commands = []\n",
    "\n",
    "readfiledir = './'\n",
    "\n",
    "# subprocess.run(['ulimit', '-n', '4096'])\n",
    "\n",
    "with open('star-command.sh','w+') as f:\n",
    "    \n",
    "    f.write('# set limit\\n')\n",
    "    f.write('ulimit -n 4096\\n')\n",
    "    \n",
    "    for input1, input2 in zip(args_1,args_2):\n",
    "\n",
    "        # create folders\n",
    "\n",
    "    #     output_folder = 'STAROUTPUT/' + input1[0:10]\n",
    "    #     path = os.path.join(parent_dir, output_folder)\n",
    "    #     try: \n",
    "    #         os.mkdir(path) \n",
    "    #     except OSError as error: \n",
    "    #         print(error)     \n",
    "\n",
    "        command = template.format(*[readfiledir+input1, readfiledir+input2, output_folder])\n",
    "        f.write('\\n\\n# id:'+ input1[0:10] + '\\n\\n')\n",
    "        f.write(command)\n",
    "#         subprocess.run(command.split(' '))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'STAR --runThreadN 80 --genomeDir /mnt/Tank/junfan/project/test1/genome --readFilesIn ./SRR7191195_1.fastq ./SRR7191195_2.fastq --outFileNamePrefix ./STAROUTPUT/SRR7191195 --outSAMtype BAM SortedByCoordinate --sjdbGTFfile /mnt/Tank/junfan/project/test1/gencode.v33.primary_assembly.annotation.gtf --sjdbOverhang 100 --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded'"
      ]
     },
     "execution_count": 504,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Jan 24 10:41:36 ..... started STAR run\n",
      "Jan 24 10:41:36 ..... loading genome\n",
      "Jan 24 10:41:54 ..... processing annotations GTF\n",
      "Jan 24 10:42:06 ..... inserting junctions into the genome indices\n",
      "^C\n"
     ]
    }
   ],
   "source": [
    "!bash star-command.sh "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "template = 'STAR --runThreadN 10 --genomeDir ' + genomedir + ' --readFilesIn {} {} --outFileNamePrefix ./{} --outSAMtype BAM SortedByCoordinate --sjdbGTFfile '+ GTFdir + ' --sjdbOverhang 100 --twopassMode Basic --outWigType bedGraph --outWigStrand Stranded'\n",
    "\n",
    "parent_dir = \"/mnt/Tank/junfan/project/test\"\n",
    "\n",
    "processes = []\n",
    "\n",
    "for input1, input2 in zip(args_1,args_2):\n",
    "    \n",
    "    # create folders\n",
    "    \n",
    "#     output_folder = 'STAROUTPUT/' + input1[0:10]\n",
    "#     path = os.path.join(parent_dir, output_folder)\n",
    "#     try: \n",
    "#         os.mkdir(path) \n",
    "#     except OSError as error: \n",
    "#         print(error)     \n",
    "    \n",
    "    command = template.format(*[input1, input2, output_folder])\n",
    "    process = subprocess.Popen(command, shell=True)\n",
    "    processes.append(process)\n",
    "\n",
    "# Collect statuses\n",
    "output = [p.wait() for p in processes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counts to Different Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are two tools to use: **edgeR** and **DESeq2**\n",
    "```{R}\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"edgeR\")\n",
    "```\n",
    "\n",
    "In order to use edgeR, you will also need to install limma.\n",
    "```\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"limma\")\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(limma)\n",
    "library(edgeR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR --runThreadN 79 --runMode genomeGenerate --genomeDir ./ \\--genomeFastaFiles ./Homo_sapiens.GRCh38.dna.primary_assembly.fa \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR \\--runThreadN 12 --genomeDir ~/star/genome/ \\--sjdbGTFfile ~/star/Homo_sapiens.GRCh38.79.gtf --sjdbOverhang 100 \\--readFilesIn SRR7191196_1.fastq SRR7191196_2.fastq \\--readFilesCommand zcat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR --runThreadN 20 --readFilesIn tq --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix /home/fanyc/RNA-seq/STAR/23 --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "STAR --runThreadN 20 --readFilesIn SRR7191196_1.fastq SRR7191196_2.fastq --quantMode TranscriptomeSAM --outSAMtype BAM SortedByCoordinate --outFileNamePrefix ./output --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --chimSegmentMin 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasterq-dump --split-3 SRR7191190"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "process = subprocess.Popen(['echo', 'More output'],\n",
    "                     stdout=subprocess.PIPE, \n",
    "                     stderr=subprocess.PIPE)\n",
    "stdout, stderr = process.communicate()\n",
    "stdout, stderr"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
